<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>(3) Analysing Data • pavo</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="(3) Analysing Data">
<meta property="og:description" content="pavo">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pavo</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased version">2.7.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/pavo-1-overview.html">Overview</a>
    </li>
    <li>
      <a href="../articles/pavo-2-importing_processing.html">Importing and Processing Data</a>
    </li>
    <li>
      <a href="../articles/pavo-3-analysing.html">Analysing Data</a>
    </li>
    <li>
      <a href="../articles/pavo-5-alphashapes.html">Computing Colour Volumes</a>
    </li>
    <li>
      <a href="../articles/pavo-4-spectraldesc.html">Spectral Shape Descriptors</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/rmaia/pavo/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="pavo-3-analysing_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>(3) Analysing Data</h1>
                        <h4 class="author">Rafael Maia, Thomas White, Hugo Gruson, Pierre-Paul Bitton, Chad Eliason</h4>
            
            <h4 class="date">2020-12-20</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/rmaia/pavo/blob/master/vignettes/pavo-3-analysing.Rmd"><code>vignettes/pavo-3-analysing.Rmd</code></a></small>
      <div class="hidden name"><code>pavo-3-analysing.Rmd</code></div>

    </div>

    
    
<style type="text/css">
/* Put all captions in italics */
.caption {
  font-style: italic;
}
/* Auto-wrap long code line */
code.sourceCode {
  white-space: pre-wrap;
}
</style>
<p>Let’s again begin by loading the package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Load the package, and set a global random-number seed for the reproducible generation of fake data later on.</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://pavo.colrverse.com">pavo</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1612217</span><span class="op">)</span></code></pre></div>
<div id="analysing-spectral-data" class="section level1">
<h1 class="hasAnchor">
<a href="#analysing-spectral-data" class="anchor"></a>Analysing Spectral Data</h1>
<div id="spectral-dataset-description" class="section level2">
<h2 class="hasAnchor">
<a href="#spectral-dataset-description" class="anchor"></a>Spectral Dataset Description</h2>
<p>The raw spectral data used in this example are available from the package repository on <a href="https://github.com/rmaia/pavo">github</a>, located <a href="https://github.com/rmaia/pavo/tree/master/data_external/vignette">here</a>. You can download and extract it to follow the vignette exactly. Alternatively, the data are included as an RData file as part of the package installation, and so can be loaded directly (see below).</p>
<p>The data consist of reflectance spectra, obtained using Avantes equipment and software, from seven bird species: Northern Cardinal <em>Cardinalis cardinalis</em>, Wattled Jacana <em>Jacana jacana</em>, Baltimore Oriole <em>Icterus galbula</em>, Peach-fronted Parakeet <em>Aratinga aurea</em>, American Robin <em>Turdus migratorius</em>, and Sayaca Tanager <em>Thraupis sayaca</em>. Several individuals were measured (sample size varies by species), and 3 spectra were collected from each individual. However, the number of individuals measured per species is uneven and the data have additional peculiarities that should emphasize the flexibility <code>pavo</code> offers, as we’ll see below.</p>
<p>In addition, <code>pavo</code> includes three datasets that can be called with the <code>data</code> function. <code><a href="https://rdrr.io/r/utils/data.html">data(teal)</a></code>, <code><a href="https://rdrr.io/r/utils/data.html">data(sicalis)</a></code>, and <code><a href="https://rdrr.io/r/utils/data.html">data(flowers)</a></code> will all be used in this vignette. See the help files for each dataset for more information; via <code><a href="../reference/teal.html">?teal</a></code>, <code><a href="../reference/sicalis.html">?sicalis</a></code>, and <code><a href="../reference/flowers.html">?flowers</a></code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">specs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/specsdata.rds"</span>, package <span class="op">=</span> <span class="st">"pavo"</span><span class="op">)</span><span class="op">)</span>
<span class="va">mspecs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggspec.html">aggspec</a></span><span class="op">(</span><span class="va">specs</span>, by <span class="op">=</span> <span class="fl">3</span>, FUN <span class="op">=</span> <span class="va">mean</span><span class="op">)</span>
<span class="va">spp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">'\\.[0-9].*$'</span>, <span class="st">''</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">mspecs</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span>
<span class="va">sppspec</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/aggspec.html">aggspec</a></span><span class="op">(</span><span class="va">mspecs</span>, by <span class="op">=</span> <span class="va">spp</span>, FUN <span class="op">=</span> <span class="va">mean</span><span class="op">)</span>
<span class="va">spec.sm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/procspec.html">procspec</a></span><span class="op">(</span><span class="va">sppspec</span>, opt <span class="op">=</span> <span class="st">"smooth"</span>, span <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span>
<span class="co">#&gt; processing options applied:</span>
<span class="co">#&gt; smoothing spectra with a span of 0.2</span></code></pre></div>
</div>
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<p><code>pavo</code> offers two main approaches for spectral data analysis. First, colour variables can be calculated based on the shape of the reflectance spectra. By using special <code>R</code> classes for spectra data frame objects, this can easily be done using the <code>summary</code> function with an <code>rspec</code> object (see below). The function <code><a href="../reference/peakshape.html">peakshape()</a></code> also returns descriptors for individual peaks in spectral curves, as outlined below.</p>
<p>Second, reflectance spectra can be analysed by accounting for the visual system receiving the colour signal, therefore representing reflectance spectra as perceived colours. To that end, we have implemented a suite of visual models and colourspaces including; the receptor-noise limited model of model of <span class="citation">Vorobyev et al. (<a href="#ref-Vorobyev1998" role="doc-biblioref">1998</a>)</span>, <span class="citation">Endler (<a href="#ref-Endler1990" role="doc-biblioref">1990</a>)</span>’s segment classification method, flexible di- tri- and tetra-chromatic colourspaces, the Hexagon model of <span class="citation">Chittka (<a href="#ref-Chittka1992" role="doc-biblioref">1992</a>)</span>, the colour-opponent coding space of <span class="citation">Backhaus (<a href="#ref-Backhaus1991" role="doc-biblioref">1991</a>)</span>, CIELAB and CIEXYZ spaces, and the categorical model of <span class="citation">Troje (<a href="#ref-Troje1993" role="doc-biblioref">1993</a>)</span>.</p>
</div>
<div id="spectral-shape-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#spectral-shape-analysis" class="anchor"></a>Spectral Shape Analysis</h2>
<div id="colourimetric-variables" class="section level3">
<h3 class="hasAnchor">
<a href="#colourimetric-variables" class="anchor"></a>Colourimetric Variables</h3>
<p>Obtaining colourimetric variables (pertaining to hue, saturation and brightness/value) is pretty straightforward in <code>pavo</code>. Since reflectance spectra is stored in an object of class <code>rspec</code>, the <code>summary</code> function recognizes the object as such and extracts 23 variables, as outlined in <span class="citation">Montgomerie (<a href="#ref-Montgomerie2006" role="doc-biblioref">2006</a>)</span>. Though outlined in a book chapter on bird colouration, these variables are broadly applicable to any reflectance data, particularly if the taxon of interest has colour vision within the UV-human visible range.</p>
<p>The description and formulas for these variables can be found by running <code><a href="../reference/summary.rspec.html">?summary.rspec</a></code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">spec.sm</span><span class="op">)</span></code></pre></div>
<p><code>summary</code> also takes an additional argument <code>subset</code> which if changed from the default <code>FALSE</code> to <code>TRUE</code> will return only the most commonly used colourimetrics <span class="citation">(Andersson and Prager <a href="#ref-Andersson2006" role="doc-biblioref">2006</a>)</span>. <code>summary</code> can also take a vector of colour variable names, which can be used to filter the results</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">spec.sm</span>, subset <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Extract only brightness variables</span>
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">spec.sm</span>, subset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'B1'</span>, <span class="st">'B2'</span>, <span class="st">'B3'</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="peak-shape-descriptors" class="section level3">
<h3 class="hasAnchor">
<a href="#peak-shape-descriptors" class="anchor"></a>Peak Shape Descriptors</h3>
<p>Particularly in cases of reflectance spectra that have multiple discrete peaks (in which case the <code>summary</code> function will only return variables based on the tallest peak in the curve), it might be useful to obtain variables that describe individual peak’s properties. The <code><a href="../reference/peakshape.html">peakshape()</a></code> function identifies the peak location (<code>H1</code>), returns the reflectance at that point (<code>B3</code>), and identifies the wavelengths at which the reflectance is half that at the peak, calculating the wavelength bandwidth of that interval (the Full Width at Half Maximum, or <code>FWHM</code>). The function also returns the half widths, which are useful when the peaks are located near the edge of the measurement limit and half maximum reflectance can only be reliably estimated from one of its sides.</p>
<p>If this all sounds too esoteric, fear not: <code><a href="../reference/peakshape.html">peakshape()</a></code> has the option of returning plots indicating what it’s calculating. The vertical continuous red line indicates the peak location, the horizontal continuous red line indicates the half-maximum reflectance, and the distance between the dashed lines (<code>HWHM.l</code> and <code>HWHM.r</code>) is the FWHM:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/peakshape.html">peakshape</a></span><span class="op">(</span><span class="va">spec.sm</span>, plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt;         id       B3  H1 FWHM HWHM.l HWHM.r incl.min</span>
<span class="co">#&gt; 1 cardinal 52.70167 700   NA    114     NA      Yes</span>
<span class="co">#&gt; 2   jacana 53.78744 700   NA    172     NA      Yes</span>
<span class="co">#&gt; 3   oriole 54.15508 700   NA    149     NA      Yes</span>
<span class="co">#&gt; 4 parakeet 29.86504 572  126     63     63      Yes</span>
<span class="co">#&gt; 5    robin 37.85542 700   NA    107     NA      Yes</span>
<span class="co">#&gt; 6  tanager 30.48108 557  166     81     85      Yes</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="pavo-3-analysing_files/figure-html/unnamed-chunk-7-1.png" alt="Plots from `peakshape`" width="30%"><img src="pavo-3-analysing_files/figure-html/unnamed-chunk-7-2.png" alt="Plots from `peakshape`" width="30%"><img src="pavo-3-analysing_files/figure-html/unnamed-chunk-7-3.png" alt="Plots from `peakshape`" width="30%"><img src="pavo-3-analysing_files/figure-html/unnamed-chunk-7-4.png" alt="Plots from `peakshape`" width="30%"><img src="pavo-3-analysing_files/figure-html/unnamed-chunk-7-5.png" alt="Plots from `peakshape`" width="30%"><img src="pavo-3-analysing_files/figure-html/unnamed-chunk-7-6.png" alt="Plots from `peakshape`" width="30%"><p class="caption">
Plots from <code>peakshape</code>
</p>
</div>
<p>As it can be seen, the variable FWHM is meaningless if the curve doesn’t have a clear peak. Sometimes, such as in the case of the Cardinal (Above figure, first panel), there might be a peak which is not the point of maximum reflectance of the entire spectral curve. <code><a href="../reference/peakshape.html">peakshape()</a></code> also offers a <code>select</code> argument to facilitate subsetting the spectra data frame to, for example, focus on a single reflectance peak:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/peakshape.html">peakshape</a></span><span class="op">(</span><span class="va">spec.sm</span>, select <span class="op">=</span> <span class="fl">2</span>, lim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">300</span>, <span class="fl">500</span><span class="op">)</span>, plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt;         id       B3  H1 FWHM HWHM.l HWHM.r incl.min</span>
<span class="co">#&gt; 1 cardinal 17.84381 369  100     46     54      Yes</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="pavo-3-analysing_files/figure-html/unnamed-chunk-8-1.png" alt="Plot from `peakshape`, setting the wavelength limits to 300 and 500 nm" width="480"><p class="caption">
Plot from <code>peakshape</code>, setting the wavelength limits to 300 and 500 nm
</p>
</div>
</div>
</div>
<div id="visual-modelling" class="section level2">
<h2 class="hasAnchor">
<a href="#visual-modelling" class="anchor"></a>Visual Modelling</h2>
<div id="overview-1" class="section level3">
<h3 class="hasAnchor">
<a href="#overview-1" class="anchor"></a>Overview</h3>
<p>All visual and colourspace modelling and plotting options are wrapped into four main function: <code><a href="../reference/vismodel.html">vismodel()</a></code>, <code><a href="../reference/colspace.html">colspace()</a></code>, <code><a href="../reference/coldist.html">coldist()</a></code> (or its bootstrap-based variant <code><a href="../reference/bootcoldist.html">bootcoldist()</a></code>), and <code>plot</code>. As detailed below, these functions cover the basic processes common to most modelling approaches. Namely, the estimation of quantum catches, their conversion to an appropriate space, estimating the distances between samples, and visualising the results. As a brief example, a typical workflow for modelling might use:</p>
<ol style="list-style-type: decimal">
<li>
<code><a href="../reference/vismodel.html">vismodel()</a></code> to estimate photoreceptor quantum catches. The assumptions of visual models may differ dramatically, so be sure to select options that are appropriate for your intended use.</li>
<li>
<code><a href="../reference/colspace.html">colspace()</a></code> to convert the results of <code><a href="../reference/vismodel.html">vismodel()</a></code> (or user-provided quantum catches) into points in a given colourspace, specified with the <code>space</code> argument. If no <code>space</code> argument is provided, <code><a href="../reference/colspace.html">colspace()</a></code> will automatically select the di-, tri- or tetrahedral colourspace, depending on the nature of the input data. The result of <code><a href="../reference/colspace.html">colspace()</a></code> will be an object of class <code>colspace</code> (that inherits from <code>data.frame</code>), and will contain the location of stimuli in the selected space along with any associated colour variables.</li>
<li>
<code><a href="../reference/coldist.html">coldist()</a></code> or <code><a href="../reference/bootcoldist.html">bootcoldist()</a></code> to estimate the colour distances between points in a colourspace (if the input is from <code>colspace</code>), or the ‘perceptual’, noise-weighted distances in the receptor-noise limited model (if the input is from <code><a href="../reference/vismodel.html">vismodel()</a></code> where <code>relative = FALSE</code>).</li>
<li>
<code>plot</code> the output. <code>plot</code> will automatically select the appropriate visualisation based on the input <code>colspace</code> object, and will also accept various graphical parameters depending on the colourspace (see <code><a href="../reference/plot.colspace.html">?plot.colspace</a></code> and links therein for details).</li>
</ol>
</div>
<div id="accessing-and-plotting-in-built-spectral-data" class="section level3">
<h3 class="hasAnchor">
<a href="#accessing-and-plotting-in-built-spectral-data" class="anchor"></a>Accessing and Plotting In-built Spectral Data</h3>
<p>All in-built spectral data in <code>pavo</code> can be easily retrieved and/or plotted via the <code><a href="../reference/sensdata.html">sensdata()</a></code> function. These data can be visualised directly by including <code>plot = TRUE</code>, or assigned to an object for subsequent use, as in:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">musca_sense</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sensdata.html">sensdata</a></span><span class="op">(</span>visual <span class="op">=</span> <span class="st">"musca"</span>, achromatic <span class="op">=</span> <span class="st">"md.r1"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">musca_sense</span><span class="op">)</span>
<span class="co">#&gt;    wl    musca.u     musca.s     musca.m     musca.l       md.r1</span>
<span class="co">#&gt; 1 300 0.01148992 0.006726680 0.001533875 0.001524705 0.001713692</span>
<span class="co">#&gt; 2 301 0.01152860 0.006749516 0.001535080 0.001534918 0.001731120</span>
<span class="co">#&gt; 3 302 0.01156790 0.006773283 0.001536451 0.001545578 0.001751792</span>
<span class="co">#&gt; 4 303 0.01160839 0.006798036 0.001537982 0.001556695 0.001775960</span>
<span class="co">#&gt; 5 304 0.01165060 0.006823827 0.001539667 0.001568283 0.001803871</span>
<span class="co">#&gt; 6 305 0.01169510 0.006850712 0.001541499 0.001580352 0.001835777</span></code></pre></div>
</div>
<div id="visual-phenotypes" class="section level3">
<h3 class="hasAnchor">
<a href="#visual-phenotypes" class="anchor"></a>Visual Phenotypes</h3>
<p><code>pavo</code> contains numerous di-, tri- and tetrachromatic visual systems, to accompany the suite of new models. The full complement of included systems are accessible via the <code><a href="../reference/vismodel.html">vismodel()</a></code> argument <code>visual</code>, and include:</p>
<table class="table">
<caption>Built-in visual phenotypes available in pavo</caption>
<colgroup>
<col width="10%">
<col width="89%">
</colgroup>
<thead><tr class="header">
<th align="left">phenotype</th>
<th align="left">description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">avg.uv</td>
<td align="left">average ultraviolet-sensitive avian (tetrachromat)</td>
</tr>
<tr class="even">
<td align="left">avg.v</td>
<td align="left">average violet-sensitive avian (tetrachromat)</td>
</tr>
<tr class="odd">
<td align="left">bluetit</td>
<td align="left">The blue tit <em>Cyanistes caeruleus</em> (tetrachromat)</td>
</tr>
<tr class="even">
<td align="left">star</td>
<td align="left">The starling <em>Sturnus vulgaris</em> (tetrachromat)</td>
</tr>
<tr class="odd">
<td align="left">pfowl</td>
<td align="left">The peafowl <em>Pavo cristatus</em> (tetrachromat)</td>
</tr>
<tr class="even">
<td align="left">apis</td>
<td align="left">The honeybee <em>Apis mellifera</em> (trichromat)</td>
</tr>
<tr class="odd">
<td align="left">ctenophorus</td>
<td align="left">The ornate dragon lizard <em>Ctenophorus ornatus</em> (trichromat)</td>
</tr>
<tr class="even">
<td align="left">canis</td>
<td align="left">The canid <em>Canis familiaris</em> (dichromat)</td>
</tr>
<tr class="odd">
<td align="left">musca</td>
<td align="left">The housefly <em>Musca domestica</em> (tetrachromat)</td>
</tr>
<tr class="even">
<td align="left">cie2</td>
<td align="left">2-degree colour matching functions for CIE models of human colour vision (trichromat)</td>
</tr>
<tr class="odd">
<td align="left">cie10</td>
<td align="left">10-degree colour matching functions for CIE models of human colour vision (trichromat)</td>
</tr>
<tr class="even">
<td align="left">segment</td>
<td align="left">A generic ‘viewer’ with broad sensitivities for use in the segment analysis of Endler (1990) (tetrachromat)</td>
</tr>
<tr class="odd">
<td align="left">habronattus</td>
<td align="left">The jumping spider <em>Habronattus pyrrithrix</em> (trichromat)</td>
</tr>
<tr class="even">
<td align="left">rhinecanthus</td>
<td align="left">The triggerfish <em>Rhinecanthus aculeatus</em> (trichromat)</td>
</tr>
</tbody>
</table>
</div>
<div id="estimating-quantum-catch" class="section level3">
<h3 class="hasAnchor">
<a href="#estimating-quantum-catch" class="anchor"></a>Estimating Quantum Catch</h3>
<p>Numerous models have been developed to understand how colours are perceived and discriminated by an individual’s or species’ visual system <span class="citation">(described in detail in Endler <a href="#ref-Endler1990" role="doc-biblioref">1990</a>; Renoult, Kelber, and Schaefer <a href="#ref-Renoult2017" role="doc-biblioref">2017</a>; Vorobyev et al. <a href="#ref-Vorobyev1998" role="doc-biblioref">1998</a>)</span>. In essence, these models take into account the receptor sensitivity of the different receptors that make the visual system in question and quantify how a given colour would stimulate those receptors individually, and their combined effect on the perception of colour. These models also have an important component of assuming and interpreting the chromatic component of colour (hue and saturation) to be processed independently of the achromatic (brightness, or luminance) component. This provides a flexible framework allowing for a tiered model construction, in which information on aspects such as different illumination sources, backgrounds, and visual systems can be considered and compared.</p>
<p>To apply any such model, we first need to quantify receptor excitation and then consider how the signal is being processed, while possibly considering the relative density of different cones and the noise-to-signal ratio.</p>
<p>To quantify the stimulation of cones by the emitted colour, we will use the <code>pavo</code> function <code><a href="../reference/vismodel.html">vismodel()</a></code>. This function takes an <code>rspec</code> dataframe as a minimal input, and the user can either select from the available options or input their own data for the additional arguments in the function:</p>
<ul>
<li>
<code>visual</code>: the visual system to be used. Available inbuilt options are detailed above, or the user may include their own dataframe, with the first column being the wavelength range and the following columns being the absorbance at each wavelength for each cone type (see below for an example).</li>
<li>
<code>achromatic</code>: Either a receptor’s sensitivity data (available options include the blue tit, chicken, and starling double-cones, and the housefly’s R1-6 receptor), which can also be user-defined as above; or the longest-wavelength receptor, the sum of the two longest-wavelength receptors, or the sum of all receptors can be used. Alternatively, <code>none</code> can be specified for no achromatic stimulation calculation.</li>
<li>
<code>illum</code>: The illuminant being considered. By default, it considers an ideal white illuminant, but implemented options are a blue sky, standard daylight, and forest shade illuminants. A vector of same length as the wavelength range being considered can also be used as the input.</li>
<li>
<code>trans</code>: Models of the effects of light transmission (e.g. through noisy environments or ocular filters). The argument defaults to <code>ideal</code> (i.e. no effect), though users can also use the built-in options of <code>bluetit</code> or <code>blackbird</code> to model the ocular transmission of blue tits/blackbirds, or specify a user-defined vector containing transmission spectra.</li>
<li>
<code>qcatch</code>: This argument determines what photon catch data should be returned
<ul>
<li>
<code>Qi</code>: The receptor quantum catches, calculated for receptor <span class="math inline">\(i\)</span> as: <span class="math display">\[Q_i = \int_\lambda{R_i(\lambda)S(\lambda)I(\lambda)d\lambda}\]</span> Where <span class="math inline">\(\lambda\)</span> denotes the wavelength, <span class="math inline">\(R_i(\lambda)\)</span> the spectral sensitivity of receptor <span class="math inline">\(i\)</span>, <span class="math inline">\(S(\lambda)\)</span> the reflectance spectrum of the colour, and <span class="math inline">\(I(\lambda)\)</span> the illuminant spectrum.</li>
<li>
<code>fi</code>: The receptor quantum catches transformed according to Fechner’s law, in which the signal of the receptor is proportional to the logarithm of the quantum catch i.e. <span class="math inline">\(f_i = \ln(Q_i)\)</span>
</li>
<li>
<code>Ei</code>: the hyperbolic transform (a simplification of the Michaelis–Menten photoreceptor equation), where <span class="math display">\[E_i = \frac{Q_i}{Q_i + 1}\]</span>.</li>
</ul>
</li>
<li>
<code>bkg</code>: The background being considered. By default, it considers an idealized background (i.e. wavelength-independent influence of the background on colour). A vector of same length as the wavelength range being considered can also be used as the input.</li>
<li>
<code>vonkries</code>: a logical argument which determines if the von Kries transformation (which normalizes receptor quantum catches to the background, thus accounting for receptor adaptation) is to be applied (defaults to <code>FALSE</code>). If <code>TRUE</code>, <span class="math inline">\(Q_i\)</span> is multiplied by a constant <span class="math inline">\(k\)</span>, which describes the von Kries transformation: <span class="math display">\[k_i = \frac{1}{\int_\lambda R_i(\lambda)S^b(\lambda)I(\lambda)d\lambda}\]</span> Where <span class="math inline">\(S^b\)</span> denotes the reflectance spectra of the background.</li>
<li>
<code>scale</code>: This argument defines how the illuminant should be scaled. The scale of the illuminant is critical of receptor noise models in which the signal intensity influences the noise (see Receptor noise section, below). Illuminant curves should be in units of <span class="math inline">\(\mu mol.s^{-1}.m^{-2}\)</span> in order to yield physiologically meaningful results. (Some software return illuminant information values in <span class="math inline">\(\mu Watt.cm^{-2}\)</span>, and must be converted to <span class="math inline">\(\mu mol.s^{-1}.m^{-2}\)</span>. This can be done by using the <code><a href="../reference/irrad2flux.html">irrad2flux()</a></code> and <code><a href="../reference/irrad2flux.html">flux2irrad()</a></code> functions.) Therefore, if the user-specified illuminant curves are <em>not</em> in these units (i.e. are measured proportional to a white standard, for example), the <code>scale</code> parameter can be used as a multiplier to yield curves that are at least a reasonable approximation of the illuminant value. Commonly used values are <strong>500</strong> for dim conditions and <strong>10,000</strong> for bright conditions.</li>
<li>
<code>relative</code>: If <code>TRUE</code>, it will make the cone stimulations relative to their sum. This is appropriate for colourspace models such as the avian tetrahedral colourspace <span class="citation">(Goldsmith <a href="#ref-Goldsmith1990" role="doc-biblioref">1990</a>; Stoddard and Prum <a href="#ref-Stoddard2008" role="doc-biblioref">2008</a>)</span>. For the photon catch and neural noise model, it is important to set <code>relative = FALSE</code>.</li>
</ul>
<p>All visual models begin with the estimation of receptor quantum catches. The requirements of models may differ significantly of course, so be sure to consult the function documentation and original publications. For this example, we will use the average reflectance of the different species to calculate the raw stimulation of retinal cones, considering the avian average UV visual system, a standard daylight illumination, and an idealized background.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vismod1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vismodel.html">vismodel</a></span><span class="op">(</span><span class="va">sppspec</span>,
  visual <span class="op">=</span> <span class="st">"avg.uv"</span>, achromatic <span class="op">=</span> <span class="st">"bt.dc"</span>,
  illum <span class="op">=</span> <span class="st">"D65"</span>, relative <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span></code></pre></div>
<p>Since there are multiple parameters that can be used to customize the output of <code><a href="../reference/vismodel.html">vismodel()</a></code>, as detailed above, for convenience these can be returned by using <code>summary</code> in a <code>vismodel</code> object:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">vismod1</span><span class="op">)</span>
<span class="co">#&gt; visual model options:</span>
<span class="co">#&gt;  * Quantal catch: Qi </span>
<span class="co">#&gt;  * Visual system, chromatic: avg.uv </span>
<span class="co">#&gt;  * Visual system, achromatic: bt.dc </span>
<span class="co">#&gt;  * Illuminant: D65, scale = 1 (von Kries colour correction not applied) </span>
<span class="co">#&gt;  * Background: ideal </span>
<span class="co">#&gt;  * Transmission: ideal </span>
<span class="co">#&gt;  * Relative: FALSE</span>
<span class="co">#&gt;        u                 s                 m                l         </span>
<span class="co">#&gt;  Min.   :0.02166   Min.   :0.04348   Min.   :0.1150   Min.   :0.1661  </span>
<span class="co">#&gt;  1st Qu.:0.03166   1st Qu.:0.07075   1st Qu.:0.1541   1st Qu.:0.1763  </span>
<span class="co">#&gt;  Median :0.03980   Median :0.07479   Median :0.2344   Median :0.2165  </span>
<span class="co">#&gt;  Mean   :0.04689   Mean   :0.10302   Mean   :0.2054   Mean   :0.2420  </span>
<span class="co">#&gt;  3rd Qu.:0.06160   3rd Qu.:0.14749   3rd Qu.:0.2482   3rd Qu.:0.2990  </span>
<span class="co">#&gt;  Max.   :0.08229   Max.   :0.18393   Max.   :0.2682   Max.   :0.3638  </span>
<span class="co">#&gt;       lum        </span>
<span class="co">#&gt;  Min.   :0.1292  </span>
<span class="co">#&gt;  1st Qu.:0.1663  </span>
<span class="co">#&gt;  Median :0.1969  </span>
<span class="co">#&gt;  Mean   :0.1926  </span>
<span class="co">#&gt;  3rd Qu.:0.2257  </span>
<span class="co">#&gt;  Max.   :0.2411</span></code></pre></div>
<p>We can visualise what <code><a href="../reference/vismodel.html">vismodel()</a></code> is doing when estimating quantum catches by comparing the reflectance spectra to the estimates they are generating:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">oldpar</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>no.readonly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">6</span><span class="op">)</span>, oma <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">3</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/layout.html">layout</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">1</span>, <span class="fl">4</span> , <span class="fl">3</span>, <span class="fl">6</span> , <span class="fl">5</span><span class="op">)</span>,
             <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">3</span> , <span class="fl">3</span>, <span class="fl">5</span> , <span class="fl">5</span><span class="op">)</span>, 
             <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">8</span>, <span class="fl">7</span>, <span class="fl">10</span>, <span class="fl">9</span>, <span class="fl">12</span>, <span class="fl">11</span><span class="op">)</span>,
             <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">7</span>, <span class="fl">7</span>, <span class="fl">9</span> , <span class="fl">9</span>, <span class="fl">11</span>, <span class="fl">11</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="va">sppspecol</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spec2rgb.html">spec2rgb</a></span><span class="op">(</span><span class="va">sppspec</span><span class="op">)</span>

<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">sppspec</span>, select <span class="op">=</span> <span class="va">i</span><span class="op">+</span><span class="fl">1</span>, col <span class="op">=</span> <span class="va">sppspecol</span>, lwd <span class="op">=</span> <span class="fl">3</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">100</span><span class="op">)</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">4.1</span>, <span class="fl">2.5</span>, <span class="fl">2.5</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">vismod1</span><span class="op">[</span><span class="va">i</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span>, yaxt <span class="op">=</span> <span class="st">"n"</span>, col <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span>
<span class="op">}</span>

<span class="fu"><a href="https://rdrr.io/r/graphics/mtext.html">mtext</a></span><span class="op">(</span><span class="st">"Wavelength (nm)"</span>, side <span class="op">=</span> <span class="fl">1</span>, outer <span class="op">=</span> <span class="cn">TRUE</span>, line <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/mtext.html">mtext</a></span><span class="op">(</span><span class="st">"Reflectance (%)"</span>, side <span class="op">=</span> <span class="fl">2</span>, outer <span class="op">=</span> <span class="cn">TRUE</span>, line <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span><span class="va">oldpar</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="pavo-3-analysing_files/figure-html/unnamed-chunk-13-1.png" alt="Plots of species mean reflectance curves with corresponding relative usml cone stimulations (insets)." width="816"><p class="caption">
Plots of species mean reflectance curves with corresponding relative usml cone stimulations (insets).
</p>
</div>
<p>As described above, <code>vismodel</code> also accepts user-defined visual systems, background and illuminants. We will illustrate this by showcasing the function <code>sensmodel</code>, which models spectral sensitivities of retinas based on their peak cone sensitivity, as described in <span class="citation">Govardovskii et al. (<a href="#ref-Govardovskii2000" role="doc-biblioref">2000</a>)</span> and <span class="citation">Hart and Vorobyev (<a href="#ref-Hart2005" role="doc-biblioref">2005</a>)</span>. <code>sensmodel</code> takes several optional arguments, but the main one is a vector containing the peak sensitivities for the cones being modelled. Let’s model an idealized dichromat visual system, with cones peaking in sensitivity at 350 and 650 nm:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">idealizeddichromat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sensmodel.html">sensmodel</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">350</span>, <span class="fl">650</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">idealizeddichromat</span>, col <span class="op">=</span> <span class="fu"><a href="../reference/spec2rgb.html">spec2rgb</a></span><span class="op">(</span><span class="va">idealizeddichromat</span><span class="op">)</span>, ylab <span class="op">=</span> <span class="st">"Absorbance"</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="pavo-3-analysing_files/figure-html/unnamed-chunk-14-1.png" alt="Idealized dichromat photoreceptors created using `sensmodel`." width="480"><p class="caption">
Idealized dichromat photoreceptors created using <code>sensmodel</code>.
</p>
</div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vismod.idi</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vismodel.html">vismodel</a></span><span class="op">(</span><span class="va">sppspec</span>, visual <span class="op">=</span> <span class="va">idealizeddichromat</span>, relative <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">vismod.idi</span></code></pre></div>
</div>
<div id="the-receptor-noise-model" class="section level3">
<h3 class="hasAnchor">
<a href="#the-receptor-noise-model" class="anchor"></a>The Receptor Noise Model</h3>
<p>The receptor-noise limited (RNL) model of <span class="citation">Vorobyev et al. (<a href="#ref-Vorobyev1998" role="doc-biblioref">1998</a>)</span>, <span class="citation">Vorobyev et al. (<a href="#ref-Vorobyev2001" role="doc-biblioref">2001</a>)</span> offers a basis for estimating the ‘perceptual’ distance between coloured stimuli (depending on available physiological and behavioural data), and assumes that the simultaneous discrimination of colours is fundamentally limited by photoreceptor noise. Colour distances under the RNL model can be calculated by using the inverse of the noise-to-signal ratio, known as the Weber fraction (<span class="math inline">\(w_i\)</span> for each cone <span class="math inline">\(i\)</span>). The Weber fraction can be calculated from the noise-to-signal ratio of cone <span class="math inline">\(i\)</span> (<span class="math inline">\(v_i\)</span>) and the relative number of receptor cells of type <span class="math inline">\(i\)</span> within the receptor field (<span class="math inline">\(n_i\)</span>):</p>
<p><span class="math display">\[w_i = \frac{v_i}{\sqrt{n_i}}\]</span></p>
<p><span class="math inline">\(w_i\)</span> is the value used for the noise when considering only neural noise mechanisms. Alternatively, the model can consider that the intensity of the colour signal itself contributes to the noise (photoreceptor, or quantum, noise). In this case, the noise for a receptor <span class="math inline">\(i\)</span> is calculated as:</p>
<p><span class="math display">\[w_i = \sqrt{\frac{v_i^2}{\sqrt{n_i}} + \frac{2}{Q_a+Q_b}}\]</span></p>
<p>where <span class="math inline">\(a\)</span> and <span class="math inline">\(b\)</span> refer to the two colour signals being compared. Note that when the values of <span class="math inline">\(Q_a\)</span> and <span class="math inline">\(Q_b\)</span> are very high, the second portion of the equation tends to zero, and the both formulas should yield similar results. Hence, it is important that the quantum catch are calculated in the appropriate illuminant scale, as described above.</p>
<p>Colour distances are obtained by weighting the Euclidean distance of the photoreceptor quantum catches by the Weber fraction of the cones (<span class="math inline">\(\Delta S\)</span>). These measurements are in units of Just Noticeable Differences (JNDs), where distances over a certain threshold (usually 1) are considered to be discernible under the conditions considered (e.g., backgrounds, illumination). The equations used in these calculations are:</p>
<p>For dichromats: <span class="math display">\[\Delta S = \sqrt{\frac{(\Delta f_1 - \Delta f_2)^2}{w_1^2+w_2^2}}\]</span></p>
<p>For trichromats: <span class="math display">\[\Delta S = \sqrt{\frac{w_1^2(\Delta f_3 - \Delta f_2)^2 + w_2^2(\Delta f_3 - \Delta f_1)^2 +
w_3^2(\Delta f_1 - \Delta f_2)^2 }{ (w_1w_2)^2 + (w_1w_3)^2 + (w_2w_3)^2 }}\]</span></p>
<p>For tetrachromats: <span class="math display">\[\Delta S =
\sqrt{(w_1w_2)^2(\Delta f_4 - \Delta f_3)^2 + (w_1w_3)^2(\Delta f_4 - \Delta f_2)^2 +
(w_1w_4)^2(\Delta f_3 - \Delta f_2)^2 + \\ (w_2w_3)^2(\Delta f_4 - \Delta f_1)^2 +
(w_2w_4)^2(\Delta f_3 - \Delta f_1)^2 + (w_3w_4)^2(\Delta f_2 - \Delta f_1)^2 / \\
((w_1w_2w_3)^2 + (w_1w_2w_4)^2 + (w_1w_3w_4)^2 + (w_2w_3w_4)^2)}\]</span></p>
<p>For the chromatic contrast. The achromatic contrast (<span class="math inline">\(\Delta L\)</span>) can be calculated based on the double cone or the receptor (or combination of receptors) responsible for chromatic processing by the equation:</p>
<p><span class="math display">\[\Delta L = \frac{\Delta f}{w}\]</span></p>
<div id="noise-weighted-distances-with-coldist-and-bootcoldist" class="section level4">
<h4 class="hasAnchor">
<a href="#noise-weighted-distances-with-coldist-and-bootcoldist" class="anchor"></a>Noise-weighted Distances with <code>coldist()</code> and <code>bootcoldist()</code>
</h4>
<p><code>pavo</code> implements the noise-weighted colour distance calculations in the functions <code><a href="../reference/coldist.html">coldist()</a></code> and <code><a href="../reference/bootcoldist.html">bootcoldist()</a></code>, assuming that raw receptor quantum catches are provided (via <code>relative = FALSE</code> in <code><a href="../reference/vismodel.html">vismodel()</a></code>). For the achromatic contrast, <code><a href="../reference/coldist.html">coldist()</a></code> uses <code>n4</code> to calculate <span class="math inline">\(w\)</span> for the achromatic contrast. Note that even if <span class="math inline">\(Q_i\)</span> is chosen, values are still log-transformed. This option is available in case the user wants to specify a data frame of quantum catches that was not generated by <code><a href="../reference/vismodel.html">vismodel()</a></code> as an input. In this case, the argument <code>qcatch</code> should be used to inform the function if <span class="math inline">\(Q_i\)</span> or <span class="math inline">\(f_i\)</span> values are being used (note that if the input to <code><a href="../reference/coldist.html">coldist()</a></code> is an object generated using the <code><a href="../reference/vismodel.html">vismodel()</a></code> function, this argument is ignored.) The type of noise to be calculated can be selected from the <code><a href="../reference/coldist.html">coldist()</a></code> argument <code>noise</code> (which accepts either <code>"neural"</code> or <code>"quantum"</code>).</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/coldist.html">coldist</a></span><span class="op">(</span><span class="va">vismod1</span>,
  noise <span class="op">=</span> <span class="st">"neural"</span>, achromatic <span class="op">=</span> <span class="cn">TRUE</span>, n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">4</span><span class="op">)</span>,
  weber <span class="op">=</span> <span class="fl">0.1</span>, weber.achro <span class="op">=</span> <span class="fl">0.1</span>
<span class="op">)</span>
<span class="co">#&gt;      patch1   patch2        dS        dL</span>
<span class="co">#&gt; 1  cardinal   jacana  8.414712 4.0043790</span>
<span class="co">#&gt; 2  cardinal   oriole  7.938730 3.5284929</span>
<span class="co">#&gt; 3  cardinal parakeet  7.991687 1.1100855</span>
<span class="co">#&gt; 4  cardinal    robin  5.888291 2.2326920</span>
<span class="co">#&gt; 5  cardinal  tanager 10.054691 2.7802473</span>
<span class="co">#&gt; 6    jacana   oriole 10.092518 0.4758862</span>
<span class="co">#&gt; 7    jacana parakeet  4.351668 2.8942935</span>
<span class="co">#&gt; 8    jacana    robin  3.470072 6.2370710</span>
<span class="co">#&gt; 9    jacana  tanager  5.011207 1.2241317</span>
<span class="co">#&gt; 10   oriole parakeet  8.071802 2.4184073</span>
<span class="co">#&gt; 11   oriole    robin  7.221260 5.7611848</span>
<span class="co">#&gt; 12   oriole  tanager 13.494597 0.7482455</span>
<span class="co">#&gt; 13 parakeet    robin  4.751953 3.3427775</span>
<span class="co">#&gt; 14 parakeet  tanager  5.933861 1.6701618</span>
<span class="co">#&gt; 15    robin  tanager  7.620600 5.0129393</span>
<span class="fu"><a href="../reference/coldist.html">coldist</a></span><span class="op">(</span><span class="va">vismod.idi</span>, n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, weber <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span>
<span class="co">#&gt;      patch1   patch2        dS dL</span>
<span class="co">#&gt; 1  cardinal   jacana 2.0993283 NA</span>
<span class="co">#&gt; 2  cardinal   oriole 4.6567462 NA</span>
<span class="co">#&gt; 3  cardinal parakeet 2.2575457 NA</span>
<span class="co">#&gt; 4  cardinal    robin 3.7236780 NA</span>
<span class="co">#&gt; 5  cardinal  tanager 3.5417700 NA</span>
<span class="co">#&gt; 6    jacana   oriole 2.5574178 NA</span>
<span class="co">#&gt; 7    jacana parakeet 4.3568740 NA</span>
<span class="co">#&gt; 8    jacana    robin 1.6243496 NA</span>
<span class="co">#&gt; 9    jacana  tanager 5.6410983 NA</span>
<span class="co">#&gt; 10   oriole parakeet 6.9142918 NA</span>
<span class="co">#&gt; 11   oriole    robin 0.9330682 NA</span>
<span class="co">#&gt; 12   oriole  tanager 8.1985162 NA</span>
<span class="co">#&gt; 13 parakeet    robin 5.9812236 NA</span>
<span class="co">#&gt; 14 parakeet  tanager 1.2842244 NA</span>
<span class="co">#&gt; 15    robin  tanager 7.2654480 NA</span></code></pre></div>
<p>Where <code>dS</code> is the chromatic contrast (<span class="math inline">\(\Delta S\)</span>) and <code>dL</code> is the achromatic contrast (<span class="math inline">\(\Delta L\)</span>). Note that, by default, <code>achromatic = FALSE</code>, so <code>dL</code> isn’t included in the second result (this is to maintain consistency since, in the <code><a href="../reference/vismodel.html">vismodel()</a></code> function, the <code>achromatic</code> argument defaults to <code>none</code>). As expected, values are really high under the avian colour vision, since the colours of these species are quite different and because of the enhanced discriminatory ability with four compared to two cones.</p>
<p><code><a href="../reference/coldist.html">coldist()</a></code> also has a <code>subset</code> argument, which is useful if only certain comparisons are of interest (for example, of colour patches against a background, or only comparisons among a species or body patch). <code>subset</code> can be a vector of length one or two. If only one subsetting option is passed, all comparisons against the matching argument are returned (useful in the case of comparing to a background, for example). If two values are passed, comparisons will only be made between samples that match that rule (partial string matching and regular expressions are accepted). For example, compare:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/coldist.html">coldist</a></span><span class="op">(</span><span class="va">vismod1</span>, subset <span class="op">=</span> <span class="st">'cardinal'</span><span class="op">)</span>
<span class="co">#&gt;     patch1   patch2        dS dL</span>
<span class="co">#&gt; 1 cardinal   jacana  8.414712 NA</span>
<span class="co">#&gt; 2 cardinal   oriole  7.938730 NA</span>
<span class="co">#&gt; 3 cardinal parakeet  7.991687 NA</span>
<span class="co">#&gt; 4 cardinal    robin  5.888291 NA</span>
<span class="co">#&gt; 5 cardinal  tanager 10.054691 NA</span></code></pre></div>
<p>to:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/coldist.html">coldist</a></span><span class="op">(</span><span class="va">vismod1</span>, subset <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'cardinal'</span>, <span class="st">'jacana'</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt;     patch1 patch2       dS dL</span>
<span class="co">#&gt; 1 cardinal jacana 8.414712 NA</span></code></pre></div>
<p>A bootstrap-based variant of <code><a href="../reference/coldist.html">coldist()</a></code>, described in <span class="citation">Maia and White (<a href="#ref-Maia2018" role="doc-biblioref">2018</a>)</span> as part of a more general analytical framework, is implemented via the function <code><a href="../reference/bootcoldist.html">bootcoldist()</a></code>. As the name suggests, <code><a href="../reference/bootcoldist.html">bootcoldist()</a></code> uses a bootstrap procedure to generate confidence intervals for the mean chromatic and/or achromatic distance between two or more samples of colours. This has the benefit of generating useful errors around distance estimates, which can be inspected for the inclusion of appropriate ‘threshold’, when asking common kinds of questions that relate to between-group differences (e.g. animals-vs-backgrounds, sexual dichromatism, colour polymorphism). The function’s arguments are shared <code><a href="../reference/coldist.html">coldist()</a></code> save for a few extra requirements. The <code>by</code> argument is used to specify a grouping variable that defines the groups to be compared, while <code>boot.n</code> and <code>alpha</code> control the number of bootstrap replicates and the confidence level for the confidence interval, respectively. The latter two have sensible defaults, but the <code>by</code> argument is required.</p>
<p>Here we run present a quick example that tests for colour differences across three body regions (breast, crown, and throat) of the finch <em>Sicalis citrina</em>, given seven measures of each region from different males.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># Load the data</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">sicalis</span><span class="op">)</span>

<span class="co"># Construct a model using an avian viewer</span>
<span class="va">sicmod</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vismodel.html">vismodel</a></span><span class="op">(</span><span class="va">sicalis</span>, visual <span class="op">=</span> <span class="st">'avg.uv'</span>, relative <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co"># Create a grouping variable to group by body region,</span>
<span class="co"># informed by the spectral data names</span>
<span class="va">regions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/substr.html">substr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">sicmod</span><span class="op">)</span>, <span class="fl">6</span>, <span class="fl">6</span><span class="op">)</span>

<span class="co"># Estimate bootstrapped colour-distances</span>
<span class="va">sicdist</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bootcoldist.html">bootcoldist</a></span><span class="op">(</span><span class="va">sicmod</span>, by <span class="op">=</span> <span class="va">regions</span>, n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">4</span><span class="op">)</span>, weber <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span>

<span class="co"># Take a look at the resulting pairwise estimates </span>
<span class="va">sicdist</span>
<span class="co">#&gt;       dS.mean    dS.lwr    dS.upr</span>
<span class="co">#&gt; B-C  9.253097 5.7015179 14.013095</span>
<span class="co">#&gt; B-T  3.483528 0.3728945  9.509223</span>
<span class="co">#&gt; C-T 12.221038 8.1736179 16.899949</span></code></pre></div>
<p>And plot the results for good measure. We can see that the pairwise comparisons of the chest, breast, and throat are quite chromatically distinct, but the breast and throat are much more similar in colour, perhaps to the point of being indistinguishable.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">sicdist</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>, 
     ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">20</span><span class="op">)</span>, 
     pch <span class="op">=</span> <span class="fl">21</span>, 
     bg <span class="op">=</span> <span class="fl">1</span>, 
     cex <span class="op">=</span> <span class="fl">2</span>, 
     xaxt <span class="op">=</span> <span class="st">'n'</span>, 
     xlab <span class="op">=</span> <span class="st">'Centroid comparison'</span>, 
     ylab <span class="op">=</span> <span class="st">'Chromatic contrast (dS)'</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/axis.html">axis</a></span><span class="op">(</span><span class="fl">1</span>, at <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">sicdist</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/segments.html">segments</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="va">sicdist</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="va">sicdist</span><span class="op">[</span>, <span class="fl">3</span><span class="op">]</span>, lwd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>  <span class="co"># Add CI's</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>h <span class="op">=</span> <span class="fl">1</span>, lty <span class="op">=</span> <span class="fl">3</span>, lwd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>  <span class="co"># Add a 'threshold' line at dS = 1</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="pavo-3-analysing_files/figure-html/unnamed-chunk-20-1.png" alt="The bootstrapped colour distances between the throat, chest, and breast regions of male Sicalis citrina." width="480"><p class="caption">
The bootstrapped colour distances between the throat, chest, and breast regions of male Sicalis citrina.
</p>
</div>
</div>
<div id="converting-receptor-noise-corrected-distances-to-cartesian-coordinates" class="section level4">
<h4 class="hasAnchor">
<a href="#converting-receptor-noise-corrected-distances-to-cartesian-coordinates" class="anchor"></a>Converting receptor noise-corrected distances to Cartesian coordinates</h4>
<p>You can convert distances in JND back to Cartesian position coordinates with the function <code><a href="../reference/jnd2xyz.html">jnd2xyz()</a></code>. Note that the actual position of these points in the XYZ space is arbitrary; therefore the function allows you to rotate the data so that, for example, the vector leading to the long-wavelength cone aligns with the X axis:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fakedata1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">vapply</a></span><span class="op">(</span>
  <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">500</span>, by <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>,
  <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span>
      <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">dnorm</a></span><span class="op">(</span><span class="fl">300</span><span class="op">:</span><span class="fl">700</span>, <span class="va">x</span>, <span class="fl">30</span><span class="op">)</span>,
      <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">dnorm</a></span><span class="op">(</span><span class="fl">300</span><span class="op">:</span><span class="fl">700</span>, <span class="va">x</span> <span class="op">+</span> <span class="fl">400</span>, <span class="fl">30</span><span class="op">)</span>
    <span class="op">)</span><span class="op">)</span>,
    <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">numeric</a></span><span class="op">(</span><span class="fl">401</span><span class="op">)</span>
<span class="op">)</span>

<span class="co"># Creating idealized specs with varying saturation</span>
<span class="va">fakedata2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">vapply</a></span><span class="op">(</span>
  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">500</span>, <span class="fl">300</span>, <span class="fl">150</span>, <span class="fl">105</span>, <span class="fl">75</span>, <span class="fl">55</span>, <span class="fl">40</span>, <span class="fl">30</span><span class="op">)</span>,
  <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">dnorm</a></span><span class="op">(</span><span class="fl">300</span><span class="op">:</span><span class="fl">700</span>, <span class="fl">550</span>, <span class="va">x</span><span class="op">)</span>,
  <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">numeric</a></span><span class="op">(</span><span class="fl">401</span><span class="op">)</span>
<span class="op">)</span>

<span class="va">fakedata1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as.rspec.html">as.rspec</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>wl <span class="op">=</span> <span class="fl">300</span><span class="op">:</span><span class="fl">700</span>, <span class="va">fakedata1</span><span class="op">)</span><span class="op">)</span>
<span class="va">fakedata1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/procspec.html">procspec</a></span><span class="op">(</span><span class="va">fakedata1</span>, <span class="st">"max"</span><span class="op">)</span>
<span class="va">fakedata2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as.rspec.html">as.rspec</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>wl <span class="op">=</span> <span class="fl">300</span><span class="op">:</span><span class="fl">700</span>, <span class="va">fakedata2</span><span class="op">)</span><span class="op">)</span>
<span class="va">fakedata2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/procspec.html">procspec</a></span><span class="op">(</span><span class="va">fakedata2</span>, <span class="st">"sum"</span><span class="op">)</span>
<span class="va">fakedata2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/procspec.html">procspec</a></span><span class="op">(</span><span class="va">fakedata2</span>, <span class="st">"min"</span><span class="op">)</span>

<span class="co"># Converting reflectance to percentage</span>
<span class="va">fakedata1</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">fakedata1</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span> <span class="op">*</span> <span class="fl">100</span>
<span class="va">fakedata2</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">fakedata2</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">fakedata2</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span>

<span class="co"># Combining and converting to rspec</span>
<span class="va">fakedata.c</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>wl <span class="op">=</span> <span class="fl">300</span><span class="op">:</span><span class="fl">700</span>, <span class="va">fakedata1</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span>, <span class="va">fakedata2</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>
<span class="va">fakedata.c</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as.rspec.html">as.rspec</a></span><span class="op">(</span><span class="va">fakedata.c</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Visual model and colour distances</span>
<span class="va">fakedata.vm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vismodel.html">vismodel</a></span><span class="op">(</span><span class="va">fakedata.c</span>, relative <span class="op">=</span> <span class="cn">FALSE</span>, achromatic <span class="op">=</span> <span class="st">'all'</span><span class="op">)</span>
<span class="va">fakedata.cd</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/coldist.html">coldist</a></span><span class="op">(</span><span class="va">fakedata.vm</span>,
  noise <span class="op">=</span> <span class="st">"neural"</span>, n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">4</span><span class="op">)</span>,
  weber <span class="op">=</span> <span class="fl">0.1</span>, achromatic <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span>

<span class="co"># Converting to Cartesian coordinates</span>
<span class="va">fakedata.cc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/jnd2xyz.html">jnd2xyz</a></span><span class="op">(</span><span class="va">fakedata.cd</span>, ref1 <span class="op">=</span> <span class="st">"l"</span>, axis1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>, ref2 <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">fakedata.cc</span><span class="op">)</span>
<span class="co">#&gt;             x          y         z      lum</span>
<span class="co">#&gt; X1 -26.165302   3.692783 -17.71718 1.430788</span>
<span class="co">#&gt; X2 -11.212399  -2.085407 -27.97937 2.315951</span>
<span class="co">#&gt; X3   3.076291  -7.441594 -37.76761 3.621349</span>
<span class="co">#&gt; X4  16.523287 -12.475488 -46.33606 4.341465</span>
<span class="co">#&gt; X5  25.812150 -17.848152 -40.71621 4.444027</span>
<span class="co">#&gt; X6  31.912643 -23.368560 -24.79660 4.046863</span></code></pre></div>
<p>which you can then plot as well:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">fakedata.cc</span>, theta <span class="op">=</span> <span class="fl">55</span>, phi <span class="op">=</span> <span class="fl">25</span>, col <span class="op">=</span> <span class="fu"><a href="../reference/spec2rgb.html">spec2rgb</a></span><span class="op">(</span><span class="va">fakedata.c</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="pavo-3-analysing_files/figure-html/unnamed-chunk-23-1.png" alt="Spectral data in a receptor noise-corrected colourspace" width="384"><p class="caption">
Spectral data in a receptor noise-corrected colourspace
</p>
</div>
<p>The axes in this colourspace are in JND units. For more information on these functions, see <code><a href="../reference/jnd2xyz.html">?jnd2xyz</a></code>, <code><a href="../reference/jndrot.html">?jndrot</a></code> and <code><a href="../reference/jndplot.html">?jndplot</a></code>.</p>
</div>
</div>
<div id="colourspaces" class="section level3">
<h3 class="hasAnchor">
<a href="#colourspaces" class="anchor"></a>Colourspaces</h3>
<p>Another general, flexible representation of stimuli is the colour space model. In such approaches, photon catches are expressed in relative values (so that the quantum catches of all receptors involved in chromatic discrimination sum to 1). The maximum stimulation of each cone <span class="math inline">\(n\)</span> is placed at the vertex of a <span class="math inline">\((n-1)\)</span>-dimensional polygon that encompasses all theoretical colours that can be perceived by that visual system. For the avian visual system comprised of 4 cones, for example, all colours can be placed somewhere in the volume of a tetrahedron, in which each of the four vertices represents the maximum stimulation of that particular cone type.</p>
<p>Though these models do not account for receptor noise (and thus do not allow an estimate of JNDs), they presents several advantages. First, they make for a very intuitive representation of colour points accounting for attributes of the colour vision of the signal receiver. Second, they allow for the calculation of several interesting variables that represent colour. For example, hue can be estimated from the angle of the point relative to the xy plane (blue-green-red) and the z axis (UV); saturation can be estimated as the distance of the point from the achromatic centre.</p>
<p><code><a href="https://rdrr.io/r/utils/data.html">data(flowers)</a></code> is a new example dataset consisting of reflectance spectra from 36 Australian angiosperm species, which we’ll use to illustrate for the following colourspace modelling.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">flowers</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">flowers</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt;    wl Goodenia_heterophylla Goodenia_geniculata Goodenia_gracilis</span>
<span class="co">#&gt; 1 300             1.7426387            1.918962         0.3638354</span>
<span class="co">#&gt; 2 301             1.5724849            1.872966         0.3501921</span>
<span class="co">#&gt; 3 302             1.4099808            1.828019         0.3366520</span>
<span class="co">#&gt; 4 303             1.2547709            1.784152         0.3231911</span>
<span class="co">#&gt; 5 304             1.1064997            1.741398         0.3097854</span>
<span class="co">#&gt; 6 305             0.9648117            1.699792         0.2964107</span></code></pre></div>
<div id="di--tri--and-tetrachromatic-colourspaces" class="section level4">
<h4 class="hasAnchor">
<a href="#di--tri--and-tetrachromatic-colourspaces" class="anchor"></a>Di-, Tri-, and Tetrachromatic Colourspaces</h4>
<p><code>pavo</code> has extensive modelling and visualisation capabilities for generic di-, tri-, and tetra-chromatic spaces, uniting these approaches in a cohesive workflow. As with most colourspace models, we first estimate relative quantum catches with various assumptions by using the <code><a href="../reference/vismodel.html">vismodel()</a></code> function, before converting each set of values to a location in colourspace by using the <code>space</code> argument in <code><a href="../reference/colspace.html">colspace()</a></code> (the function can also be set to try to detect the dimensionality of the colourspace automatically). For di- tri- and tetrachromatic spaces, <code><a href="../reference/colspace.html">colspace()</a></code> calculates the coordinates of stimuli as:</p>
<p><strong>Dichromats:</strong> <span class="math display">\[x = \frac{1}{\sqrt{2}}(Q_l - Q_s)\]</span></p>
<p><strong>Trichromats:</strong> <span class="math display">\[x = \frac{1}{\sqrt{2}}(Q_l - Q_m)\]</span> <span class="math display">\[y = \frac{\sqrt{2}}{\sqrt{3}}(Q_s - \frac{Q_l + Q_m}{2})\]</span></p>
<p><strong>Tetrachromats:</strong> <span class="math display">\[x = \frac{1}{\sqrt{2}}(Q_l - Q_m)\]</span> <span class="math display">\[y = \frac{\sqrt{2}}{\sqrt{3}}(Q_s - \frac{Q_l + Q_m}{2})\]</span> <span class="math display">\[z = \frac{\sqrt{3}}{2}(Q_u - \frac{Q_l + Q_m + Q_s}{3})\]</span></p>
<p>Where Q<sub>u</sub>, Q<sub>s</sub>, Q<sub>m</sub>, and Q<sub>l</sub> refer to quantum catch estimates for UV-, short, medium-, and long-wavelength photoreceptors, respectively.</p>
<p>For a dichromatic example, we can model our floral reflectance data using the visual system of the domestic dog <em>Canis familiaris</em>, which has two cones with maximal sensitivity near 440 and 560 nm.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vis.flowers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vismodel.html">vismodel</a></span><span class="op">(</span><span class="va">flowers</span>, visual <span class="op">=</span> <span class="st">'canis'</span><span class="op">)</span>

<span class="va">di.flowers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/colspace.html">colspace</a></span><span class="op">(</span><span class="va">vis.flowers</span>, space <span class="op">=</span> <span class="st">'di'</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">di.flowers</span><span class="op">)</span>
<span class="co">#&gt;                                s         l           x      r.vec lum</span>
<span class="co">#&gt; Goodenia_heterophylla 0.52954393 0.4704561 -0.04178143 0.04178143  NA</span>
<span class="co">#&gt; Goodenia_geniculata   0.25430150 0.7456985  0.34747015 0.34747015  NA</span>
<span class="co">#&gt; Goodenia_gracilis     0.01747832 0.9825217  0.68238870 0.68238870  NA</span>
<span class="co">#&gt; Xyris_operculata      0.39433933 0.6056607  0.14942675 0.14942675  NA</span>
<span class="co">#&gt; Eucalyptus_sp         0.40628552 0.5937145  0.13253229 0.13253229  NA</span>
<span class="co">#&gt; Faradaya_splendida    0.23166580 0.7683342  0.37948187 0.37948187  NA</span></code></pre></div>
<p>The output contains values for the relative stimulation of short- and long-wavelength sensitive photoreceptors associated with each flower, along with its single coordinate in dichromatic space and its r.vector (distance from the origin). To visualise where these points lie, we can simply plot them on a segment.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">di.flowers</span>, col <span class="op">=</span> <span class="fu"><a href="../reference/spec2rgb.html">spec2rgb</a></span><span class="op">(</span><span class="va">flowers</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="pavo-3-analysing_files/figure-html/unnamed-chunk-26-1.png" alt="Flowers in a dichromatic colourspace, as modelled according to a canid visual system." width="384"><p class="caption">
Flowers in a dichromatic colourspace, as modelled according to a canid visual system.
</p>
</div>
<p>For our trichromatic viewer we can use the honeybee <em>Apis mellifera</em>, one of the most significant and widespread pollinators. We’ll also transform our quantum catches according to Fechner’s law by specifying <code>qcatch = 'fi'</code>, and will model photoreceptor stimulation under bright conditions by scaling our illuminant with the <code>scale</code> argument.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vis.flowers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vismodel.html">vismodel</a></span><span class="op">(</span><span class="va">flowers</span>, visual <span class="op">=</span> <span class="st">'apis'</span>, qcatch <span class="op">=</span> <span class="st">'fi'</span>, scale <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span>

<span class="va">tri.flowers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/colspace.html">colspace</a></span><span class="op">(</span><span class="va">vis.flowers</span>, space <span class="op">=</span> <span class="st">'tri'</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">tri.flowers</span><span class="op">)</span>
<span class="co">#&gt;                               s         m         l            x           y</span>
<span class="co">#&gt; Goodenia_heterophylla 0.2882383 0.3587254 0.3530363 -0.004022828 -0.05522995</span>
<span class="co">#&gt; Goodenia_geniculata   0.2475326 0.3549832 0.3974842  0.030052720 -0.10508396</span>
<span class="co">#&gt; Goodenia_gracilis     0.1373117 0.3125304 0.5501579  0.168028001 -0.24007647</span>
<span class="co">#&gt; Xyris_operculata      0.2395035 0.3729820 0.3875145  0.010275990 -0.11491764</span>
<span class="co">#&gt; Eucalyptus_sp         0.2760091 0.3548096 0.3691813  0.010162377 -0.07020755</span>
<span class="co">#&gt; Faradaya_splendida    0.2654951 0.3431266 0.3913783  0.034119147 -0.08308452</span>
<span class="co">#&gt;                          h.theta      r.vec lum</span>
<span class="co">#&gt; Goodenia_heterophylla -1.6435057 0.05537626  NA</span>
<span class="co">#&gt; Goodenia_geniculata   -1.2922439 0.10929687  NA</span>
<span class="co">#&gt; Goodenia_gracilis     -0.9601417 0.29303604  NA</span>
<span class="co">#&gt; Xyris_operculata      -1.4816131 0.11537616  NA</span>
<span class="co">#&gt; Eucalyptus_sp         -1.4270471 0.07093922  NA</span>
<span class="co">#&gt; Faradaya_splendida    -1.1811377 0.08981733  NA</span></code></pre></div>
<p>As in the case of our dichromat, the output contains relative photoreceptor stimulations, coordinates in the Maxwell triangle, as well as the ‘hue angle’ <code>h.theta</code> and distance from the origin (<code>r.vec</code>).</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">tri.flowers</span>, pch <span class="op">=</span> <span class="fl">21</span>, bg <span class="op">=</span> <span class="fu"><a href="../reference/spec2rgb.html">spec2rgb</a></span><span class="op">(</span><span class="va">flowers</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="pavo-3-analysing_files/figure-html/unnamed-chunk-28-1.png" alt="Floral reflectance in a Maxwell triangle, considering a honeybee visual system." width="384"><p class="caption">
Floral reflectance in a Maxwell triangle, considering a honeybee visual system.
</p>
</div>
<p>Finally, we’ll draw on the blue tit’s visual system to model our floral reflectance spectra in a tetrahedral space, again using log-transformed quantum catches and assuming bright viewing conditions.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vis.flowers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vismodel.html">vismodel</a></span><span class="op">(</span><span class="va">flowers</span>, visual <span class="op">=</span> <span class="st">"bluetit"</span>, qcatch <span class="op">=</span> <span class="st">"fi"</span>, scale <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span>

<span class="va">tetra.flowers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/colspace.html">colspace</a></span><span class="op">(</span><span class="va">vis.flowers</span>, space <span class="op">=</span> <span class="st">"tcs"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">tetra.flowers</span><span class="op">)</span>
<span class="co">#&gt;                               u         s         m         l         u.r</span>
<span class="co">#&gt; Goodenia_heterophylla 0.2274310 0.2659669 0.2524170 0.2541851 -0.02256897</span>
<span class="co">#&gt; Goodenia_geniculata   0.1625182 0.2721621 0.2828016 0.2825182 -0.08748184</span>
<span class="co">#&gt; Goodenia_gracilis     0.0428552 0.2443091 0.3504414 0.3623943 -0.20714480</span>
<span class="co">#&gt; Xyris_operculata      0.1766234 0.2709785 0.2642637 0.2881344 -0.07337659</span>
<span class="co">#&gt; Eucalyptus_sp         0.2118481 0.2609468 0.2637023 0.2635028 -0.03815190</span>
<span class="co">#&gt; Faradaya_splendida    0.1844183 0.2557660 0.2786393 0.2811764 -0.06558169</span>
<span class="co">#&gt;                                s.r         m.r         l.r            x</span>
<span class="co">#&gt; Goodenia_heterophylla  0.015966893 0.002417012 0.004185066 -0.007214866</span>
<span class="co">#&gt; Goodenia_geniculata    0.022162064 0.032801599 0.032518179  0.006341799</span>
<span class="co">#&gt; Goodenia_gracilis     -0.005690920 0.100441373 0.112394348  0.072312163</span>
<span class="co">#&gt; Xyris_operculata       0.020978454 0.014263689 0.038134446  0.010505857</span>
<span class="co">#&gt; Eucalyptus_sp          0.010946788 0.013702317 0.013502794  0.001565228</span>
<span class="co">#&gt; Faradaya_splendida     0.005765955 0.028639328 0.031176408  0.015560661</span>
<span class="co">#&gt;                                  y           z    h.theta     h.phi      r.vec</span>
<span class="co">#&gt; Goodenia_heterophylla -0.005415708 -0.02256897 -2.4976873 -1.190529 0.02430520</span>
<span class="co">#&gt; Goodenia_geniculata    0.003861848 -0.08748184  0.5469755 -1.486123 0.08779638</span>
<span class="co">#&gt; Goodenia_gracilis      0.033297418 -0.20714480  0.4315247 -1.203879 0.22191605</span>
<span class="co">#&gt; Xyris_operculata      -0.010813615 -0.07337659 -0.7998327 -1.368146 0.07490949</span>
<span class="co">#&gt; Eucalyptus_sp          0.001044768 -0.03815190  0.5885699 -1.521510 0.03819828</span>
<span class="co">#&gt; Faradaya_splendida     0.007189965 -0.06558169  0.4328380 -1.315140 0.06778487</span>
<span class="co">#&gt;                           r.max r.achieved lum</span>
<span class="co">#&gt; Goodenia_heterophylla 0.2692325 0.09027589  NA</span>
<span class="co">#&gt; Goodenia_geniculata   0.2508989 0.34992737  NA</span>
<span class="co">#&gt; Goodenia_gracilis     0.2678272 0.82857920  NA</span>
<span class="co">#&gt; Xyris_operculata      0.2552227 0.29350636  NA</span>
<span class="co">#&gt; Eucalyptus_sp         0.2503039 0.15260760  NA</span>
<span class="co">#&gt; Faradaya_splendida    0.2583986 0.26232676  NA</span></code></pre></div>
<p>Tetrahedral data (via <code><a href="../reference/colspace.html">colspace(space = 'tcs')</a></code>) may be plotted in a standard <em>static</em> tetrahedron by using <code><a href="https://rdrr.io/r/graphics/plot.default.html">plot()</a></code>, or can be visualised as part of an <em>interactive</em> tetrahedron by using <code><a href="../reference/tcsplot.html">tcsplot()</a></code> with the accessory functions <code><a href="../reference/tcsplot.html">tcspoints()</a></code> and <code><a href="../reference/tcsplot.html">tcsvol()</a></code> for adding points and convex hulls, respectively. As with other colourspace plots there are a number of associated graphical options, though the <code>theta</code> and <code>phi</code> arguments are particularly useful in this case, as they control the orientation (in degrees) of the tetrahedron in the xy and yz planes, respectively.</p>
<p>When plotting the tetrahedral colourspace, one can also force perspective by changing the size of points relative to their distance from the plane of observation, using the arguments <code>perspective = TRUE</code> and controlling the size range with the <code>range</code> argument. Several other options control the appearance of this plot, you can check these using <code><a href="../reference/plot.colspace.html">?plot.colspace</a></code> or <code><a href="../reference/tetraplot.html">?tetraplot</a></code></p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">tetra.flowers</span>, pch <span class="op">=</span> <span class="fl">21</span>, bg <span class="op">=</span> <span class="fu"><a href="../reference/spec2rgb.html">spec2rgb</a></span><span class="op">(</span><span class="va">flowers</span><span class="op">)</span>, perspective <span class="op">=</span> <span class="cn">TRUE</span>, range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, cex <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="pavo-3-analysing_files/figure-html/unnamed-chunk-30-1.png" alt="Flowers in a tetrahedral colourspace modelled using the visual phenotype of the blue tit. Point size is used to force perspective" width="480"><p class="caption">
Flowers in a tetrahedral colourspace modelled using the visual phenotype of the blue tit. Point size is used to force perspective
</p>
</div>
<p>Two additional functions may help with tetrahedral colourspace plotting:</p>
<ul>
<li>
<code><a href="../reference/axistetra.html">axistetra()</a></code> function can be used to draw arrows showing the direction and magnitude of distortion of x, y and z in the tetrahedral plot.</li>
<li>
<code><a href="../reference/legendtetra.html">legendtetra()</a></code> allows you to add a legend to a plot.</li>
</ul>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">tetra.flowers</span>, pch <span class="op">=</span> <span class="fl">21</span>, bg <span class="op">=</span> <span class="fu"><a href="../reference/spec2rgb.html">spec2rgb</a></span><span class="op">(</span><span class="va">flowers</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/axistetra.html">axistetra</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">0</span>, y <span class="op">=</span> <span class="fl">1.8</span><span class="op">)</span>
<span class="co">#&gt; Warning in arrows(xarr$x[1], xarr$y[1], xarr$x[2], xarr$y[2], length =</span>
<span class="co">#&gt; arrowhead, : zero-length arrow is of indeterminate angle and so skipped</span>
<span class="co">#&gt; Warning in arrows(yarr$x[1], yarr$y[1], yarr$x[2], yarr$y[2], length =</span>
<span class="co">#&gt; arrowhead, : zero-length arrow is of indeterminate angle and so skipped</span>
<span class="co">#&gt; Warning in arrows(zarr$x[1], zarr$y[1], zarr$x[2], zarr$y[2], length =</span>
<span class="co">#&gt; arrowhead, : zero-length arrow is of indeterminate angle and so skipped</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">tetra.flowers</span>, theta <span class="op">=</span> <span class="fl">110</span>, phi <span class="op">=</span> <span class="fl">10</span>, pch <span class="op">=</span> <span class="fl">21</span>, bg <span class="op">=</span> <span class="fu"><a href="../reference/spec2rgb.html">spec2rgb</a></span><span class="op">(</span><span class="va">flowers</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/axistetra.html">axistetra</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">0</span>, y <span class="op">=</span> <span class="fl">1.8</span><span class="op">)</span>
<span class="co">#&gt; Warning in arrows(xarr$x[1], xarr$y[1], xarr$x[2], xarr$y[2], length =</span>
<span class="co">#&gt; arrowhead, : zero-length arrow is of indeterminate angle and so skipped</span>
<span class="co">#&gt; Warning in arrows(yarr$x[1], yarr$y[1], yarr$x[2], yarr$y[2], length =</span>
<span class="co">#&gt; arrowhead, : zero-length arrow is of indeterminate angle and so skipped</span>
<span class="co">#&gt; Warning in arrows(zarr$x[1], zarr$y[1], zarr$x[2], zarr$y[2], length =</span>
<span class="co">#&gt; arrowhead, : zero-length arrow is of indeterminate angle and so skipped</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="pavo-3-analysing_files/figure-html/unnamed-chunk-31-1.png" alt="Flowers in a tetrahedral colourspace, with varied orientations and perspectives, modelled using the visual phenotype of the blue tit." width="45%"><img src="pavo-3-analysing_files/figure-html/unnamed-chunk-31-2.png" alt="Flowers in a tetrahedral colourspace, with varied orientations and perspectives, modelled using the visual phenotype of the blue tit." width="45%"><p class="caption">
Flowers in a tetrahedral colourspace, with varied orientations and perspectives, modelled using the visual phenotype of the blue tit.
</p>
</div>
<p>For tetrahedral models, another plotting option available is <code>projplot</code>, which projects colour points in the surface of a sphere encompassing the tetrahedron. This plot is particularly useful to see differences in hue.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/projplot.html">projplot</a></span><span class="op">(</span><span class="va">tetra.flowers</span>, pch <span class="op">=</span> <span class="fl">20</span>, col <span class="op">=</span> <span class="fu"><a href="../reference/spec2rgb.html">spec2rgb</a></span><span class="op">(</span><span class="va">flowers</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="pavo-3-analysing_files/figure-html/unnamed-chunk-32-1.png" alt="Projection plot from a tetrahedral colour space." width="480"><p class="caption">
Projection plot from a tetrahedral colour space.
</p>
</div>
<p>Finally, a useful function for tri- and tetrahedral models is <code><a href="../reference/voloverlap.html">voloverlap()</a></code>, which calculates the overlap in tri- or tetrahedral colour volume between two sets of points. This can be useful to explore whether different species occupy similar (overlapping) or different (non- overlapping) “sensory niches”, or to test for mimetism, dichromatism, etc. <span class="citation">(Stoddard and Stevens <a href="#ref-Stoddard2011" role="doc-biblioref">2011</a>)</span>. To show this function, we will use the <code>sicalis</code> dataset, which includes measurements from the crown (C), throat (T) and breast (B) of seven stripe-tailed yellow finches (<em>Sicalis citrina</em>).</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">sicalis</span><span class="op">)</span></code></pre></div>
<p>We will use this dataset to test for the overlap between the volume determined by the measurements of those body parts from multiple individuals in the tetrahedral colourspace (note the option <code>plot</code> for plotting of the volumes):</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tcs.sicalis.C</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="fu"><a href="../reference/colspace.html">colspace</a></span><span class="op">(</span><span class="fu"><a href="../reference/vismodel.html">vismodel</a></span><span class="op">(</span><span class="va">sicalis</span><span class="op">)</span><span class="op">)</span>, <span class="st">"C"</span><span class="op">)</span>
<span class="va">tcs.sicalis.T</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="fu"><a href="../reference/colspace.html">colspace</a></span><span class="op">(</span><span class="fu"><a href="../reference/vismodel.html">vismodel</a></span><span class="op">(</span><span class="va">sicalis</span><span class="op">)</span><span class="op">)</span>, <span class="st">"T"</span><span class="op">)</span>
<span class="va">tcs.sicalis.B</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="fu"><a href="../reference/colspace.html">colspace</a></span><span class="op">(</span><span class="fu"><a href="../reference/vismodel.html">vismodel</a></span><span class="op">(</span><span class="va">sicalis</span><span class="op">)</span><span class="op">)</span>, <span class="st">"B"</span><span class="op">)</span>
<span class="fu"><a href="../reference/voloverlap.html">voloverlap</a></span><span class="op">(</span><span class="va">tcs.sicalis.T</span>, <span class="va">tcs.sicalis.B</span>, plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt;           vol1         vol2   overlapvol vsmallest      vboth</span>
<span class="co">#&gt; 1 5.183721e-06 6.281511e-06 6.904074e-07 0.1331876 0.06407598</span>
<span class="fu"><a href="../reference/voloverlap.html">voloverlap</a></span><span class="op">(</span><span class="va">tcs.sicalis.T</span>, <span class="va">tcs.sicalis.C</span>, plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt;           vol1         vol2 overlapvol vsmallest vboth</span>
<span class="co">#&gt; 1 5.183721e-06 4.739152e-06          0         0     0</span></code></pre></div>
<p><img src="pavo-3-analysing_files/figure-html/unnamed-chunk-34-1.png" width="40%" style="display: block; margin: auto;"><img src="pavo-3-analysing_files/figure-html/unnamed-chunk-34-2.png" width="40%" style="display: block; margin: auto;"></p>
<p>The function <code><a href="../reference/voloverlap.html">voloverlap()</a></code> gives the volume (<span class="math inline">\(V\)</span>) of the convex hull delimited by the overlap between the two original volumes, and two proportions are calculated from that: <span class="math display">\[\text{vsmallest} = V_{overlap} / \min(V_A, V_B)\]</span> and <span class="math display">\[\text{vboth} = V_{overlap} / (V_A + V_B)\]</span>. Thus, if one of the volumes is entirely contained in the other, <code>vsmallest</code> will equal 1.</p>
<p>So we can clearly see that there is overlap between the throat and breast colours (of about 6%), but not between the throat and the crown colours (Figures above).</p>
<p><strong>Summary variables for groups of points</strong></p>
<p>Another advantage of colourspace models is that they allow for the calculation of useful summary statistics of groups of points, such as the centroid of the points, the total volume occupied, the mean and variance of hue span and the mean saturation. In <code>pavo</code>, the result of a <code><a href="../reference/colspace.html">colspace()</a></code> call is an object of class <code>colspace</code>, and thus these summary statistics can be calculated simply by calling <code>summary</code>. Note that the <code>summary</code> call can also take a <code>by</code> vector of group identities, so that the variables are calculated for each group separately:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">tetra.flowers</span><span class="op">)</span>
<span class="co">#&gt; Colorspace &amp; visual model options:</span>
<span class="co">#&gt;  * Colorspace: tcs </span>
<span class="co">#&gt;  * Quantal catch: fi </span>
<span class="co">#&gt;  * Visual system, chromatic: bluetit </span>
<span class="co">#&gt;  * Visual system, achromatic: none </span>
<span class="co">#&gt;  * Illuminant: ideal, scale = 10000 (von Kries colour correction not applied) </span>
<span class="co">#&gt;  * Background: ideal </span>
<span class="co">#&gt;  * Relative: TRUE </span>
<span class="co">#&gt;  * Max possible chromatic volume: NA</span>
<span class="co">#&gt; 'avalue' automatically set to 1.9861e-01</span>
<span class="co">#&gt;            centroid.u centroid.s centroid.m centroid.l        c.vol   rel.c.vol</span>
<span class="co">#&gt; all.points  0.1948478  0.2554421  0.2704411   0.279269 0.0002510915 0.001159742</span>
<span class="co">#&gt;             colspan.m   colspan.v huedisp.m huedisp.v   mean.ra    max.ra</span>
<span class="co">#&gt; all.points 0.05706825 0.001682224 0.6413439 0.3157949 0.2428508 0.8285792</span>
<span class="co">#&gt;                   a.vol</span>
<span class="co">#&gt; all.points 0.0001723672</span></code></pre></div>
</div>
<div id="the-colour-hexagon" class="section level4">
<h4 class="hasAnchor">
<a href="#the-colour-hexagon" class="anchor"></a>The Colour Hexagon</h4>
<p>The hexagon colour space of <span class="citation">Chittka (<a href="#ref-Chittka1992" role="doc-biblioref">1992</a>)</span> is a generalised colour-opponent model of hymenopteran vision that has found extremely broad use, particularly in studies of bee-flower interactions. It’s also often broadly applied across hymenopteran species, because the photopigments underlying trichromatic vision in Hymenoptera appear to be quite conserved <span class="citation">(Briscoe and Chittka <a href="#ref-Briscoe2001" role="doc-biblioref">2001</a>)</span>. What’s particularly useful is that colour distances within the hexagon have been extensively validated against behaviour, and thus offer a relatively reliable measure of perceptual distance.</p>
<p>In the hexagon, photoreceptor quantum catches are typically hyperbolically transformed (and <code>pavo</code> will return a warning if the transform is not selected), and vonkries correction is often used used to model photoreceptor adaptation to a vegetation background. This can all now be specified in <code><a href="../reference/vismodel.html">vismodel()</a></code>. including the optional use of a ‘green’ vegetation background. Note that although this is a colourspace model, we specific <code>relative = FALSE</code> to return unnormalised quantum catches, as required for the model.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vis.flowers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vismodel.html">vismodel</a></span><span class="op">(</span><span class="va">flowers</span>, visual <span class="op">=</span> <span class="st">'apis'</span>, qcatch <span class="op">=</span> <span class="st">'Ei'</span>, relative <span class="op">=</span> <span class="cn">FALSE</span>, vonkries <span class="op">=</span> <span class="cn">TRUE</span>, achromatic <span class="op">=</span> <span class="st">'l'</span>, bkg <span class="op">=</span> <span class="st">'green'</span><span class="op">)</span></code></pre></div>
<p>We can then apply the hexagon model in <code>colspace</code>, which will convert our photoreceptor ‘excitation values’ to coordinates in the hexagon according to:</p>
<p><span class="math display">\[x = \frac{\sqrt{3}}{2(E_g + E_{uv})}\]</span></p>
<p><span class="math display">\[y = E_b - 0.5(E_{uv} + E_g)\]</span></p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hex.flowers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/colspace.html">colspace</a></span><span class="op">(</span><span class="va">vis.flowers</span>, space <span class="op">=</span> <span class="st">'hexagon'</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">hex.flowers</span><span class="op">)</span>
<span class="co">#&gt;                                s         m         l         x          y</span>
<span class="co">#&gt; Goodenia_heterophylla 0.56572097 0.8237141 0.7053057 0.1208839  0.1882007</span>
<span class="co">#&gt; Goodenia_geniculata   0.35761236 0.8176153 0.8670134 0.4411542  0.2053024</span>
<span class="co">#&gt; Goodenia_gracilis     0.01888788 0.1622766 0.7810589 0.6600594 -0.2376968</span>
<span class="co">#&gt; Xyris_operculata      0.21080752 0.7345122 0.6796464 0.4060263  0.2892853</span>
<span class="co">#&gt; Eucalyptus_sp         0.55622758 0.8515289 0.8208038 0.2291297  0.1630132</span>
<span class="co">#&gt; Faradaya_splendida    0.45855056 0.7828905 0.8565895 0.3447118  0.1253205</span>
<span class="co">#&gt;                         h.theta     r.vec sec.fine sec.coarse       lum</span>
<span class="co">#&gt; Goodenia_heterophylla  32.71320 0.2236793       30  bluegreen 0.1680153</span>
<span class="co">#&gt; Goodenia_geniculata    65.04388 0.4865862       60  bluegreen 0.3548825</span>
<span class="co">#&gt; Goodenia_gracilis     109.80467 0.7015541      100      green 0.2313678</span>
<span class="co">#&gt; Xyris_operculata       54.53092 0.4985412       50  bluegreen 0.1518319</span>
<span class="co">#&gt; Eucalyptus_sp          54.57024 0.2812005       50  bluegreen 0.2787543</span>
<span class="co">#&gt; Faradaya_splendida     70.02119 0.3667853       70  bluegreen 0.3351008</span></code></pre></div>
<p>Again, the output includes the photoreceptor excitation values for short- medium- and long-wave sensitive photoreceptors, x and y coordinates, and measures of hue and saturation for each stimulus. The hexagon model also outputs two additional measures of of subjective ‘bee-hue’; <code>sec.fine</code> and <code>sec.coarse</code>. <code>sec.fine</code> describes the location of stimuli within one of 36 ‘hue sectors’ that are specified by radially dissecting the hexagon in 10-degree increments. <code>sec.coarse</code> follows a similar principle, though here the hexagon is divided into only five ‘bee-hue’ sectors: UV, UV-blue, blue, blue-green, green, and UV-green. These can easily be visualised by specifying <code>sectors = 'coarse</code> or <code>sectors = 'fine'</code> in a call to <code>plot</code> after modelling.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">hex.flowers</span>, sectors <span class="op">=</span> <span class="st">'coarse'</span>, pch <span class="op">=</span> <span class="fl">21</span>, bg <span class="op">=</span> <span class="fu"><a href="../reference/spec2rgb.html">spec2rgb</a></span><span class="op">(</span><span class="va">flowers</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="pavo-3-analysing_files/figure-html/unnamed-chunk-38-1.png" alt="Flowers as modelled in the hymenopteran colour hexagon of Chittka (1992), overlain with coarse bee-hue sectors." width="480"><p class="caption">
Flowers as modelled in the hymenopteran colour hexagon of Chittka (1992), overlain with coarse bee-hue sectors.
</p>
</div>
</div>
<div id="the-colour-opponent-coding-coc-space" class="section level4">
<h4 class="hasAnchor">
<a href="#the-colour-opponent-coding-coc-space" class="anchor"></a>The Colour Opponent Coding (COC) Space</h4>
<p>The colour opponent coding (coc) space is an earlier hymenopteran visual model <span class="citation">(Backhaus <a href="#ref-Backhaus1991" role="doc-biblioref">1991</a>)</span> which, although now seldom used in favour of the hexagon, may prove useful for comparative work. While the initial estimation of photoreceptor excitation is similar to that in the hexagon, the coc subsequently specifies <code>A</code> and <code>B</code> coordinates based on empirically-derived weights for the output from each photoreceptor:</p>
<p><span class="math display">\[A = -9.86E_u + 7.70E_b + 2.16E_g\]</span> <span class="math display">\[B = -5.17E_u + 20.25E_b - 15.08E_g\]</span></p>
<p>Where <span class="math inline">\(E_i\)</span> is the excitation value (hyperbolic-transformed quantum catch) in photoreceptor <span class="math inline">\(i\)</span>.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vis.flowers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vismodel.html">vismodel</a></span><span class="op">(</span><span class="va">flowers</span>, visual <span class="op">=</span> <span class="st">"apis"</span>, qcatch <span class="op">=</span> <span class="st">"Ei"</span>, relative <span class="op">=</span> <span class="cn">FALSE</span>, vonkries <span class="op">=</span> <span class="cn">TRUE</span>, bkg <span class="op">=</span> <span class="st">"green"</span><span class="op">)</span>

<span class="va">coc.flowers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/colspace.html">colspace</a></span><span class="op">(</span><span class="va">vis.flowers</span>, space <span class="op">=</span> <span class="st">"coc"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">coc.flowers</span><span class="op">)</span>
<span class="co">#&gt;                                s         m         l        x          y</span>
<span class="co">#&gt; Goodenia_heterophylla 0.56572097 0.8237141 0.7053057 2.288050  3.1194224</span>
<span class="co">#&gt; Goodenia_geniculata   0.35761236 0.8176153 0.8670134 4.642329  1.6332926</span>
<span class="co">#&gt; Goodenia_gracilis     0.01888788 0.1622766 0.7810589 2.750382 -8.5899171</span>
<span class="co">#&gt; Xyris_operculata      0.21080752 0.7345122 0.6796464 5.045218  3.5349309</span>
<span class="co">#&gt; Eucalyptus_sp         0.55622758 0.8515289 0.8208038 2.845304  1.9900420</span>
<span class="co">#&gt; Faradaya_splendida    0.45855056 0.7828905 0.8565895 3.357182  0.5654568</span>
<span class="co">#&gt;                           r.vec lum</span>
<span class="co">#&gt; Goodenia_heterophylla  5.407473  NA</span>
<span class="co">#&gt; Goodenia_geniculata    6.275622  NA</span>
<span class="co">#&gt; Goodenia_gracilis     11.340299  NA</span>
<span class="co">#&gt; Xyris_operculata       8.580149  NA</span>
<span class="co">#&gt; Eucalyptus_sp          4.835346  NA</span>
<span class="co">#&gt; Faradaya_splendida     3.922638  NA</span></code></pre></div>
<p>The A and B coordinates are designated x and y in the output of coc for consistency, and while the model includes a measure of saturation in <code>r.vec</code>, it contains no associated measure of hue.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">coc.flowers</span>, pch <span class="op">=</span> <span class="fl">21</span>, bg <span class="op">=</span> <span class="fu"><a href="../reference/spec2rgb.html">spec2rgb</a></span><span class="op">(</span><span class="va">flowers</span><span class="op">)</span>, yaxt <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="pavo-3-analysing_files/figure-html/unnamed-chunk-40-1.png" alt="Flowers in the colour-opponent-coding space of Backhaus (1991), as modelling according to the honeybee." width="480"><p class="caption">
Flowers in the colour-opponent-coding space of Backhaus (1991), as modelling according to the honeybee.
</p>
</div>
</div>
<div id="cie-spaces" class="section level4">
<h4 class="hasAnchor">
<a href="#cie-spaces" class="anchor"></a>CIE Spaces</h4>
<p>The CIE (International Commission on Illumination) colourspaces are a suite of models of human colour vision and perception. <code>pavo 1.0</code> now includes two of the most commonly used: the foundational 1931 CIE XYZ space, and the more modern, perceptually calibrated CIE LAB space and its cylindrical CIE LCh transformation. Tristimulus values in XYZ space are calculated as:</p>
<p><span class="math display">\[X = k\int_{300}^{700}{R(\lambda)I(\lambda)\bar{x}(\lambda)\,d\lambda}\]</span> <span class="math display">\[Y = k\int_{300}^{700}{R(\lambda)I(\lambda)\bar{y}(\lambda)\,d\lambda}\]</span> <span class="math display">\[Z = k\int_{300}^{700}{R(\lambda)I(\lambda)\bar{z}(\lambda)\,d\lambda}\]</span></p>
<p>where <em>x</em>, <em>y</em>, and <em>z</em> are the trichromatic colour matching functions for a ‘standard colourimetric viewer’. These functions are designed to describe an average human’s chromatic response within a specified viewing arc in the fovea (to account for the uneven distribution of cones across eye). <code>pavo</code> includes both the CIE 2-degree and the modern 10-degree standard observer, which can be selected in the <code>visual</code> option in the <code>vismodel</code> function. In these equations, k is the normalising factor</p>
<p><span class="math display">\[k = \frac{100}{\int_{300}^{700}{I(\lambda)\bar{y}(\lambda)\,d\lambda}}\]</span></p>
<p>and the chromaticity coordinates of stimuli are calculated as</p>
<p><span class="math display">\[x = \frac{X}{X + Y + Z}\]</span> <span class="math display">\[y = \frac{Y}{X + Y + Z}\]</span> <span class="math display">\[z = \frac{Z}{X + Y + Z} = 1 - x - y\]</span></p>
<p>For modelling in both XYZ and LAB spaces, here we’ll use the CIE 10-degree standard observer, and assume a <code>D65</code> ‘standard daylight’ illuminant. Again, although they are colourspace models, we need to set <code>relative = FALSE</code> to return raw quantum catch estimates, <code>vonkries = TRUE</code> to account for the required normalising factor (as above), and <code>achromatic = 'none'</code> since there is no applicable way to estimate luminance in the CIE models. As with all models that have particular requirements, <code><a href="../reference/vismodel.html">vismodel()</a></code> will output warnings and/or errors if unusual or non-standard arguments are specified.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vis.flowers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vismodel.html">vismodel</a></span><span class="op">(</span><span class="va">flowers</span>, visual <span class="op">=</span> <span class="st">'cie10'</span>, illum <span class="op">=</span> <span class="st">'D65'</span>, vonkries <span class="op">=</span> <span class="cn">TRUE</span>, relative <span class="op">=</span> <span class="cn">FALSE</span>, achromatic <span class="op">=</span> <span class="st">'none'</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ciexyz.flowers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/colspace.html">colspace</a></span><span class="op">(</span><span class="va">vis.flowers</span>, space <span class="op">=</span> <span class="st">'ciexyz'</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">ciexyz.flowers</span><span class="op">)</span>
<span class="co">#&gt;                               X         Y          Z         x         y</span>
<span class="co">#&gt; Goodenia_heterophylla 0.2130116 0.2096604 0.30085962 0.2944054 0.2897736</span>
<span class="co">#&gt; Goodenia_geniculata   0.6282685 0.6725795 0.41507113 0.3661411 0.3919646</span>
<span class="co">#&gt; Goodenia_gracilis     0.4769607 0.4385423 0.01620888 0.5119187 0.4706845</span>
<span class="co">#&gt; Xyris_operculata      0.2776811 0.2281172 0.22057916 0.3822820 0.3140478</span>
<span class="co">#&gt; Eucalyptus_sp         0.4398245 0.4463310 0.40026285 0.3418985 0.3469563</span>
<span class="co">#&gt; Faradaya_splendida    0.6351234 0.6442301 0.28463920 0.4060910 0.4119138</span>
<span class="co">#&gt;                                z</span>
<span class="co">#&gt; Goodenia_heterophylla 0.41582098</span>
<span class="co">#&gt; Goodenia_geniculata   0.24189434</span>
<span class="co">#&gt; Goodenia_gracilis     0.01739688</span>
<span class="co">#&gt; Xyris_operculata      0.30367016</span>
<span class="co">#&gt; Eucalyptus_sp         0.31114517</span>
<span class="co">#&gt; Faradaya_splendida    0.18199523</span></code></pre></div>
<p>The output is simply the tristimulus values and chromaticity coordinates of stimuli, and we can visualise our results (along with a line connecting monochromatic loci, by default) by calling</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">ciexyz.flowers</span>, pch <span class="op">=</span> <span class="fl">21</span>, bg <span class="op">=</span> <span class="fu"><a href="../reference/spec2rgb.html">spec2rgb</a></span><span class="op">(</span><span class="va">flowers</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="pavo-3-analysing_files/figure-html/unnamed-chunk-43-1.png" alt="Floral reflectance in the CIEXYZ human visual model. Note that this space is not perceptually calibrated, so we cannot make inferences about the similarity or differences of colours based on their relative location." width="480"><p class="caption">
Floral reflectance in the CIEXYZ human visual model. Note that this space is not perceptually calibrated, so we cannot make inferences about the similarity or differences of colours based on their relative location.
</p>
</div>
<p>The <code>Lab</code> space is a more recent development, and is a colour-opponent model that attempts to mimic the nonlinear responses of the human eye. The <code>Lab</code> space is also calibrated (unlike the <code>XYZ</code> space) such that Euclidean distances between points represent relative perceptual distances. This means that two stimuli that are farther apart in <code>Lab</code> space should be <em>perceived</em> as more different than two closer points. As the name suggests, the dimensions of the <code>Lab</code> space are <em>L</em>ightness (i.e. subjective brightness), along with two colour-opponent dimensions designated <em>a</em> and <em>b</em>. The <code><a href="../reference/colspace.html">colspace()</a></code> function, when <code>space = cielab</code>, simply converts points from the <code>XYZ</code> model according to:</p>
<p><span class="math display">\[
L=\begin{cases}
116\left(\frac{Y}{Y_n}\right)^\frac{1}{3} &amp; \text{if } \frac{Y}{Y_n} &gt; 0.008856 \\
903.3\left(\frac{Y}{Y_n}\right) &amp; \text{if } \frac{Y}{Y_n} \leq 0.008856
\end{cases}
\]</span></p>
<p><span class="math display">\[a = 500\left(f\left(\frac{X}{X_n}\right) - f\left(\frac{Y}{Y_n}\right)\right)\]</span></p>
<p><span class="math display">\[b = 500\left(f\left(\frac{Y}{Y_n}\right) - f\left(\frac{Z}{Z_n}\right)\right)\]</span></p>
<p>where</p>
<p><span class="math display">\[
f(x)=\begin{cases}
x^\frac{1}{3} &amp; \text{if } x &gt; 0.008856 \\
7.787x + \frac{4}{29} &amp; \text{if } x \leq 0.008856
\end{cases}
\]</span></p>
<p>Here, <span class="math inline">\(X_n, Y_n, Z_n\)</span> are neutral point values to model visual adaptation, calculated as:</p>
<p><span class="math display">\[X_n = \int_{300}^{700}{R_n(\lambda)I(\lambda)\bar{x}(\lambda)\,d\lambda}\]</span> <span class="math display">\[Y_n = \int_{300}^{700}{R_n(\lambda)I(\lambda)\bar{y}(\lambda)\,d\lambda}\]</span> <span class="math display">\[Z_n = \int_{300}^{700}{R_n(\lambda)I(\lambda)\bar{z}(\lambda)\,d\lambda}\]</span></p>
<p>when <span class="math inline">\(R_n(\lambda)\)</span> is a perfect diffuse reflector (i.e. 1).</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cielab.flowers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/colspace.html">colspace</a></span><span class="op">(</span><span class="va">vis.flowers</span>, space <span class="op">=</span> <span class="st">'cielab'</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">cielab.flowers</span><span class="op">)</span>
<span class="co">#&gt;                               X         Y          Z        L         a</span>
<span class="co">#&gt; Goodenia_heterophylla 0.2130116 0.2096604 0.30085962 19.20722  5.972526</span>
<span class="co">#&gt; Goodenia_geniculata   0.6282685 0.6725795 0.41507113 61.61574 -3.784589</span>
<span class="co">#&gt; Goodenia_gracilis     0.4769607 0.4385423 0.01620888 40.17534 25.580630</span>
<span class="co">#&gt; Xyris_operculata      0.2776811 0.2281172 0.22057916 20.89807 25.632161</span>
<span class="co">#&gt; Eucalyptus_sp         0.4398245 0.4463310 0.40026285 40.88887  7.030413</span>
<span class="co">#&gt; Faradaya_splendida    0.6351234 0.6442301 0.28463920 59.01862 10.266281</span>
<span class="co">#&gt;                                b</span>
<span class="co">#&gt; Goodenia_heterophylla -11.165703</span>
<span class="co">#&gt; Goodenia_geniculata    45.142142</span>
<span class="co">#&gt; Goodenia_gracilis      66.881871</span>
<span class="co">#&gt; Xyris_operculata        3.565477</span>
<span class="co">#&gt; Eucalyptus_sp          11.585827</span>
<span class="co">#&gt; Faradaya_splendida     59.861744</span></code></pre></div>
<p>Our output now contains the tristimulus <code>XYZ</code> values, as well as their <code>Lab</code> counterparts, which are coordinates in the <code>Lab</code> space. These can also be visualised in three-dimensional <code>Lab</code> space by calling <code>plot</code>:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">cielab.flowers</span>, pch <span class="op">=</span> <span class="fl">21</span>, bg <span class="op">=</span> <span class="fu"><a href="../reference/spec2rgb.html">spec2rgb</a></span><span class="op">(</span><span class="va">flowers</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="pavo-3-analysing_files/figure-html/unnamed-chunk-45-1.png" alt="Floral reflectance spectra represented in the CIELab model of human colour sensation." width="480"><p class="caption">
Floral reflectance spectra represented in the CIELab model of human colour sensation.
</p>
</div>
</div>
<div id="categorical-fly-colourspace" class="section level4">
<h4 class="hasAnchor">
<a href="#categorical-fly-colourspace" class="anchor"></a>Categorical Fly Colourspace</h4>
<p>The categorical colour vision model of <span class="citation">Troje (<a href="#ref-Troje1993" role="doc-biblioref">1993</a>)</span> is a model of dipteran vision, based on behavioural data from the blowfly <em>Lucilia</em> sp. It assumes the involvement of all four dipteran photoreceptor classes (R7 &amp; R8 ‘pale’ and ‘yellow’ subtypes), and further posits that colour vision is based on two specific opponent mechanisms (R7p - R8p, and R7y - R8y). The model assumes that all colours are perceptually grouped into one of four colour categories, and that flies are unable to distinguish between colours that fall within the same category.</p>
<p>We’ll use the visual systems of the muscoid fly <em>Musca domestica</em>, and will begin by estimating linear (i.e. untransformed) quantum catches for each of the four photoreceptors.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vis.flowers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vismodel.html">vismodel</a></span><span class="op">(</span><span class="va">flowers</span>, qcatch <span class="op">=</span> <span class="st">'Qi'</span>, visual <span class="op">=</span> <span class="st">'musca'</span>, achromatic <span class="op">=</span> <span class="st">'none'</span>, relative <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>Our call to <code><a href="../reference/colspace.html">colspace()</a></code> will then simply estimate the location of stimuli in the categorical space as the difference in relative stimulation between ‘pale’ (R7p - R8p) and ‘yellow’ (R7y - R8y) photoreceptor pairs:</p>
<p><span class="math display">\[x = R7_p - R8_p\]</span></p>
<p><span class="math display">\[y = R7_y - R8_y\]</span></p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cat.flowers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/colspace.html">colspace</a></span><span class="op">(</span><span class="va">vis.flowers</span>, space <span class="op">=</span> <span class="st">'categorical'</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">cat.flowers</span><span class="op">)</span>
<span class="co">#&gt;                              R7p        R7y       R8p       R8y          x</span>
<span class="co">#&gt; Goodenia_heterophylla 0.05387616 0.18710662 0.4335986 0.3254187 -0.3797224</span>
<span class="co">#&gt; Goodenia_geniculata   0.02080098 0.08276030 0.3737592 0.5226795 -0.3529583</span>
<span class="co">#&gt; Goodenia_gracilis     0.01153130 0.02501417 0.1420049 0.8214496 -0.1304736</span>
<span class="co">#&gt; Xyris_operculata      0.01270353 0.12523419 0.4488377 0.4132246 -0.4361342</span>
<span class="co">#&gt; Eucalyptus_sp         0.03461949 0.14207848 0.3977208 0.4255812 -0.3631013</span>
<span class="co">#&gt; Faradaya_splendida    0.03599953 0.09188560 0.3129195 0.5591953 -0.2769200</span>
<span class="co">#&gt;                                y category     r.vec   h.theta lum</span>
<span class="co">#&gt; Goodenia_heterophylla -0.1383120     p-y- 0.4041279 -2.792284  NA</span>
<span class="co">#&gt; Goodenia_geniculata   -0.4399192     p-y- 0.5640110 -2.246954  NA</span>
<span class="co">#&gt; Goodenia_gracilis     -0.7964355     p-y- 0.8070519 -1.733176  NA</span>
<span class="co">#&gt; Xyris_operculata      -0.2879904     p-y- 0.5226390 -2.557993  NA</span>
<span class="co">#&gt; Eucalyptus_sp         -0.2835027     p-y- 0.4606695 -2.478681  NA</span>
<span class="co">#&gt; Faradaya_splendida    -0.4673097     p-y- 0.5431971 -2.105745  NA</span></code></pre></div>
<p>And it is simply the signs of these differences that define the four possible fly-colour categories (p+y+, p-y+, p+y-, p-y-), which we can see in the associated plot.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">cat.flowers</span>, pch <span class="op">=</span> <span class="fl">21</span>, bg <span class="op">=</span> <span class="fu"><a href="../reference/spec2rgb.html">spec2rgb</a></span><span class="op">(</span><span class="va">flowers</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="pavo-3-analysing_files/figure-html/unnamed-chunk-48-1.png" alt="Flowers in the categorical colourspace of Troje (1993)." width="480"><p class="caption">
Flowers in the categorical colourspace of Troje (1993).
</p>
</div>
</div>
<div id="segment-classification" class="section level4">
<h4 class="hasAnchor">
<a href="#segment-classification" class="anchor"></a>Segment Classification</h4>
<p>The segment classification analysis <span class="citation">(Endler <a href="#ref-Endler1990" role="doc-biblioref">1990</a>)</span> does not assume any particular visual system, but instead tries to classify colours in a manner that captures common properties of many vertebrate (and some invertebrate) visual systems. In essence, it breaks down the reflectance spectrum region of interest into four equally-spaced regions, measuring the relative signal along those regions. This approximates a tetrachromatic system with ultraviolet, short, medium, and long-wavelength sensitive photoreceptors.</p>
<p>Though somewhat simplistic, this model captures many of the properties of other, more complex visual models, but without many of the additional assumptions these make. It also provides results in a fairly intuitive colour space, in which the angle corresponds to hue and the distance from the centre corresponds to chroma (Figure below; in fact, variables <code>S5</code> and <code>H4</code> from <code><a href="../reference/summary.rspec.html">summary.rspec()</a></code> are calculated from these relative segments). Note that, while a segment analysis ranging from 300 or 400 nm to 700 nm corresponds quite closely to the human visual system colour wheel, any wavelength range can be analysed in this way, returning a 360° hue space delimited by the range used.</p>
<p>The segment differences or “opponents” are calculated as:</p>
<p><span class="math display">\[LM = \frac{ R_\lambda \sum_{\lambda={Q4}} R_\lambda - \sum_{\lambda={Q2}} R_\lambda }{\sum_{\lambda={min}}^{max}R_\lambda}\]</span> <span class="math display">\[MS = \frac{ R_\lambda \sum_{\lambda={Q3}} R_\lambda - \sum_{\lambda={Q1}} R_\lambda }{\sum_{\lambda={min}}^{max}R_\lambda}\]</span></p>
<p>Where <span class="math inline">\(Qi\)</span> represent the interquantile distances (e.g. <span class="math inline">\(Q1=ultraviolet\)</span>, <span class="math inline">\(Q2=blue\)</span>, <span class="math inline">\(Q3=green\)</span> and <span class="math inline">\(Q4=red\)</span>)</p>
<p>The segment classification model is obtained through the <code><a href="../reference/colspace.html">colspace()</a></code> argument <code>space = 'segment'</code>, following initial modelling with <code><a href="../reference/vismodel.html">vismodel()</a></code>. The example below uses idealized reflectance spectra to illustrate how the avian colour space defined from the segment classification maps to the human colour wheel:</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fakedata1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">vapply</a></span><span class="op">(</span>
  <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">500</span>, by <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>,
  <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span>
      <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">dnorm</a></span><span class="op">(</span><span class="fl">300</span><span class="op">:</span><span class="fl">700</span>, <span class="va">x</span>, <span class="fl">30</span><span class="op">)</span>,
      <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">dnorm</a></span><span class="op">(</span><span class="fl">300</span><span class="op">:</span><span class="fl">700</span>, <span class="va">x</span> <span class="op">+</span> <span class="fl">400</span>, <span class="fl">30</span><span class="op">)</span>
    <span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">numeric</a></span><span class="op">(</span><span class="fl">401</span><span class="op">)</span>
<span class="op">)</span>

<span class="co"># creating idealized specs with varying saturation</span>
<span class="va">fakedata2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">vapply</a></span><span class="op">(</span>
  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">500</span>, <span class="fl">300</span>, <span class="fl">150</span>, <span class="fl">105</span>, <span class="fl">75</span>, <span class="fl">55</span>, <span class="fl">40</span>, <span class="fl">30</span><span class="op">)</span>,
  <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">dnorm</a></span><span class="op">(</span><span class="fl">300</span><span class="op">:</span><span class="fl">700</span>, <span class="fl">550</span>, <span class="va">x</span><span class="op">)</span>,
  <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">numeric</a></span><span class="op">(</span><span class="fl">401</span><span class="op">)</span>
<span class="op">)</span>

<span class="va">fakedata1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as.rspec.html">as.rspec</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>wl <span class="op">=</span> <span class="fl">300</span><span class="op">:</span><span class="fl">700</span>, <span class="va">fakedata1</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; wavelengths found in column 1</span>
<span class="va">fakedata1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/procspec.html">procspec</a></span><span class="op">(</span><span class="va">fakedata1</span>, <span class="st">"max"</span><span class="op">)</span>
<span class="co">#&gt; processing options applied:</span>
<span class="co">#&gt; Scaling spectra to a maximum value of 1</span>
<span class="va">fakedata2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as.rspec.html">as.rspec</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>wl <span class="op">=</span> <span class="fl">300</span><span class="op">:</span><span class="fl">700</span>, <span class="va">fakedata2</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; wavelengths found in column 1</span>
<span class="va">fakedata2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/procspec.html">procspec</a></span><span class="op">(</span><span class="va">fakedata2</span>, <span class="st">"sum"</span><span class="op">)</span>
<span class="co">#&gt; processing options applied:</span>
<span class="co">#&gt; Scaling spectra to a total area of 1</span>
<span class="va">fakedata2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/procspec.html">procspec</a></span><span class="op">(</span><span class="va">fakedata2</span>, <span class="st">"min"</span><span class="op">)</span>
<span class="co">#&gt; processing options applied:</span>
<span class="co">#&gt; Scaling spectra to a minimum value of zero</span>

<span class="co"># converting reflectance to percentage</span>
<span class="va">fakedata1</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">fakedata1</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span> <span class="op">*</span> <span class="fl">100</span>
<span class="va">fakedata2</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">fakedata2</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">fakedata2</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span>

<span class="co"># combining and converting to rspec</span>
<span class="va">fakedata.c</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>wl <span class="op">=</span> <span class="fl">300</span><span class="op">:</span><span class="fl">700</span>, <span class="va">fakedata1</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span>, <span class="va">fakedata2</span><span class="op">[</span>, <span class="op">-</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>
<span class="va">fakedata.c</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as.rspec.html">as.rspec</a></span><span class="op">(</span><span class="va">fakedata.c</span><span class="op">)</span>
<span class="co">#&gt; wavelengths found in column 1</span>

<span class="co"># segment classification analysis</span>
<span class="va">seg.vis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vismodel.html">vismodel</a></span><span class="op">(</span><span class="va">fakedata.c</span>, visual <span class="op">=</span> <span class="st">"segment"</span>, achromatic <span class="op">=</span> <span class="st">"all"</span><span class="op">)</span>
<span class="va">seg.fdc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/colspace.html">colspace</a></span><span class="op">(</span><span class="va">seg.vis</span>, space <span class="op">=</span> <span class="st">"segment"</span><span class="op">)</span>

<span class="co"># plot results</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">seg.fdc</span>, col <span class="op">=</span> <span class="fu"><a href="../reference/spec2rgb.html">spec2rgb</a></span><span class="op">(</span><span class="va">fakedata.c</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="pavo-3-analysing_files/figure-html/unnamed-chunk-49-1.png" alt="Idealized reflectance spectra and their projection on the axes of segment classification" width="384"><p class="caption">
Idealized reflectance spectra and their projection on the axes of segment classification
</p>
</div>
</div>
<div id="colourspace-distances-with-coldist" class="section level4">
<h4 class="hasAnchor">
<a href="#colourspace-distances-with-coldist" class="anchor"></a>Colourspace Distances with <code>coldist</code>
</h4>
<p>Under the colour space framework, colour distances can be calculated simply as Euclidean distances of the relative cone stimulation data, either log-transformed or not, depending on how it was defined. However, these distances cannot be interpreted in terms of JNDs, since no receptor noise is incorporated in the model. Euclidean distances can be computed in <code>R</code> using the <code><a href="../reference/coldist.html">coldist()</a></code> function on the <code><a href="../reference/colspace.html">colspace()</a></code> output:</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/coldist.html">coldist</a></span><span class="op">(</span><span class="va">tetra.flowers</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Warning: Quantum catch are relative, distances may not be meaningful</span>
<span class="co">#&gt; Calculating unweighted Euclidean distances</span>
<span class="co">#&gt;                  patch1              patch2         dS dL</span>
<span class="co">#&gt; 1 Goodenia_heterophylla Goodenia_geniculata 0.06695922 NA</span>
<span class="co">#&gt; 2 Goodenia_heterophylla   Goodenia_gracilis 0.20467411 NA</span>
<span class="co">#&gt; 3 Goodenia_heterophylla    Xyris_operculata 0.05407934 NA</span>
<span class="co">#&gt; 4 Goodenia_heterophylla       Eucalyptus_sp 0.01901724 NA</span>
<span class="co">#&gt; 5 Goodenia_heterophylla  Faradaya_splendida 0.05027645 NA</span>
<span class="co">#&gt; 6 Goodenia_heterophylla  Gaultheria_hispida 0.03897671 NA</span></code></pre></div>
<p>Specialised colourspace models such as the colour-hexagon, coc space, CIELab and CIELCh models have their own distance measures, which are returned be default when run through <code><a href="../reference/coldist.html">coldist()</a></code>. Be sure to read the <code><a href="../reference/coldist.html">?coldist</a></code> documentation, and the original publications, to understand what is being returned. As an example, if we run the results of colour-hexagon modelling through <code><a href="../reference/coldist.html">coldist()</a></code>:</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Model flower colours according to a honeybee</span>
<span class="va">vis.flowers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vismodel.html">vismodel</a></span><span class="op">(</span><span class="va">flowers</span>, visual <span class="op">=</span> <span class="st">"apis"</span>, qcatch <span class="op">=</span> <span class="st">"Ei"</span>, relative <span class="op">=</span> <span class="cn">FALSE</span>, vonkries <span class="op">=</span> <span class="cn">TRUE</span>, achromatic <span class="op">=</span> <span class="st">"l"</span>, bkg <span class="op">=</span> <span class="st">"green"</span><span class="op">)</span>
<span class="va">hex.flowers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/colspace.html">colspace</a></span><span class="op">(</span><span class="va">vis.flowers</span>, space <span class="op">=</span> <span class="st">"hexagon"</span><span class="op">)</span>

<span class="co"># Estimate colour distances. No need to specify relative receptor densities, noise etc.,</span>
<span class="co"># which only apply in the case of receptor-noise modelling</span>
<span class="va">dist.flowers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/coldist.html">coldist</a></span><span class="op">(</span><span class="va">hex.flowers</span><span class="op">)</span>
<span class="co">#&gt; Calculating unweighted Euclidean distances</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">dist.flowers</span><span class="op">)</span>
<span class="co">#&gt;                  patch1              patch2        dS dL</span>
<span class="co">#&gt; 1 Goodenia_heterophylla Goodenia_geniculata 0.3207265 NA</span>
<span class="co">#&gt; 2 Goodenia_heterophylla   Goodenia_gracilis 0.6870945 NA</span>
<span class="co">#&gt; 3 Goodenia_heterophylla    Xyris_operculata 0.3025298 NA</span>
<span class="co">#&gt; 4 Goodenia_heterophylla       Eucalyptus_sp 0.1111375 NA</span>
<span class="co">#&gt; 5 Goodenia_heterophylla  Faradaya_splendida 0.2324927 NA</span>
<span class="co">#&gt; 6 Goodenia_heterophylla  Gaultheria_hispida 0.2280836 NA</span></code></pre></div>
<p>the chromatic contrasts <code>dS</code> and achromatic contrasts <code>dL</code> are expressed as Euclidean distances in the hexagon, known as ‘hexagon units’ <span class="citation">(Chittka <a href="#ref-Chittka1992" role="doc-biblioref">1992</a>)</span>. If we had instead used the colour-opponent coding space the units would have been city-bloc distances, while in the CIELab model the distances would be derived from the CIE colour-distance formula (2000).</p>
</div>
<div id="distances-in-n-dimensions" class="section level4">
<h4 class="hasAnchor">
<a href="#distances-in-n-dimensions" class="anchor"></a>Distances in N-Dimensions</h4>
<p>The <code><a href="../reference/coldist.html">coldist()</a></code> function is no longer limited to di-, tri- or tetrachromatic visual systems. The function has been generalised, and can now calculate colour distances for n-dimensional visual phenotypes. This means there is no limit on the dimensionality of visual systems that may be input, which may prove useful for modelling nature’s extremes (e.g. butterflies, mantis shrimp) or for simulation-based work. Naturally, since these calculations aren’t largely implemented elsewhere, we recommend caution and validation of results prior to publication.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Create an arbitrary visual phenotype with 10 photoreceptors</span>
<span class="va">fakemantisshrimp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sensmodel.html">sensmodel</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">325</span>, <span class="fl">350</span>, <span class="fl">400</span>, <span class="fl">425</span>, <span class="fl">450</span>, <span class="fl">500</span>, <span class="fl">550</span>, <span class="fl">600</span>, <span class="fl">650</span>, <span class="fl">700</span><span class="op">)</span>, beta <span class="op">=</span> <span class="cn">FALSE</span>, integrate <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co"># Convert to percentages, just to colour the plot</span>
<span class="va">fakemantisshrimp.colours</span> <span class="op">&lt;-</span> <span class="va">fakemantisshrimp</span> <span class="op">*</span> <span class="fl">100</span>
<span class="va">fakemantisshrimp.colours</span><span class="op">[</span>, <span class="st">"wl"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">fakemantisshrimp</span><span class="op">[</span>, <span class="st">"wl"</span><span class="op">]</span>

<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">fakemantisshrimp</span>, col <span class="op">=</span> <span class="fu"><a href="../reference/spec2rgb.html">spec2rgb</a></span><span class="op">(</span><span class="va">fakemantisshrimp.colours</span><span class="op">)</span>, lwd <span class="op">=</span> <span class="fl">2</span>, ylab <span class="op">=</span> <span class="st">"Absorbance"</span><span class="op">)</span>

<span class="co"># Run visual model and calculate colour distances</span>
<span class="va">vm.fms</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vismodel.html">vismodel</a></span><span class="op">(</span><span class="va">flowers</span>, visual <span class="op">=</span> <span class="va">fakemantisshrimp</span>, relative <span class="op">=</span> <span class="cn">FALSE</span>, achromatic <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="va">JND.fms</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/coldist.html">coldist</a></span><span class="op">(</span><span class="va">vm.fms</span>, n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">4</span>, <span class="fl">5</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Calculating noise-weighted Euclidean distances</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">JND.fms</span><span class="op">)</span>
<span class="co">#&gt;                  patch1              patch2        dS dL</span>
<span class="co">#&gt; 1 Goodenia_heterophylla Goodenia_geniculata 16.297557 NA</span>
<span class="co">#&gt; 2 Goodenia_heterophylla   Goodenia_gracilis 48.390309 NA</span>
<span class="co">#&gt; 3 Goodenia_heterophylla    Xyris_operculata 30.760482 NA</span>
<span class="co">#&gt; 4 Goodenia_heterophylla       Eucalyptus_sp  6.044184 NA</span>
<span class="co">#&gt; 5 Goodenia_heterophylla  Faradaya_splendida 15.901435 NA</span>
<span class="co">#&gt; 6 Goodenia_heterophylla  Gaultheria_hispida 11.188246 NA</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="pavo-3-analysing_files/figure-html/unnamed-chunk-52-1.png" alt="Visual system of a pretend mantis shrimp with 10 cones" width="384"><p class="caption">
Visual system of a pretend mantis shrimp with 10 cones
</p>
</div>
</div>
</div>
</div>
</div>
<div id="analysing-spatial-data" class="section level1">
<h1 class="hasAnchor">
<a href="#analysing-spatial-data" class="anchor"></a>Analysing Spatial Data</h1>
<p><code>pavo</code> now allows for the combined analysis of spectral data and spatial data. Note that ‘spatial’ data of course include images, but may also draw on spatially-sampled <em>spectra</em>, as explored below. Thus while much of the below deals with image-based processing and analysis, images are not necessary for many spatial analyses and indeed may perform worse than spectrally-based analyses in some cases. Further, <strong>image data should seldom be used without input from spectral measurements</strong>. Even in the simple butterfly example below, users should first validate the presence and number of discrete colours in a patch via spectral measurement and visual modelling (i.e. the ‘spectral’ information), which is then augmented by image-based clustering of these <em>a priori</em> identified patches to extract the ‘spatial’ information. Using uncalibrated images alone for inferences about colour pattern perception may well yield unreliable results, given the radical differences between human (camera) and non-human animal perception.</p>
<div id="image-based-colour-classification" class="section level2">
<h2 class="hasAnchor">
<a href="#image-based-colour-classification" class="anchor"></a>Image-based colour classification</h2>
<p>We first import our images with <code><a href="../reference/getimg.html">getimg()</a></code>, which functions in a similar manner to <code><a href="../reference/getspec.html">getspec()</a></code>. Since we are importing more than one image, we can simply point the function to the folder containing the images, and it will use parallel processing (where possible) to import any jpg, bmp, or png images in the folder. As with the raw spectra, these images are also available at the package repository [here][data-location], but are included in the package installation for convenience.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">butterflies</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getimg.html">getimg</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"testdata/images/"</span>, package <span class="op">=</span> <span class="st">'pavo'</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; 2 files found; importing images.</span></code></pre></div>
<p>The segmentation of RGB images into discrete colour-classes is an useful step on the road to several analyses, and this is carried out in <code>pavo</code> via <code><a href="../reference/classify.html">classify()</a></code>. The function currently implements k-means clustering which, and offers some flexibility in the way in which it is carried out. In the simplest case, we can specify an image or image(s) as well as the number of discrete colours in each, and allow the k-means clustering algorithm to do its work. <strong>Note that the specification of <em>k</em>, the number of discrete colours in an image, is a crucial step to be estimated <em>a priori</em> beforehand</strong>, typically through the use of visual modelling to estimate the number of discriminable colours in an ‘objective’ manner. There are a suite of possible ways in which this may be achieved, though, with one detailed below as part of example 2.</p>
<p>Since we have two images, with (arguably) four and three colours each, including the white background, we can specify this information using <code>kcols</code>, before visualising the results with <code>summary</code>. The <code>summary</code> function, with <code>plot = TRUE</code>, is also particularly useful for understanding the colour-class labels, or ID’s, assigned to each colour, as well as visualising the accuracy of the classification process, which may be enhanced by specifying strongly-contrasting ‘false’ colours for plotting through the <code>col</code> argument.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">butterflies_class</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/classify.html">classify</a></span><span class="op">(</span><span class="va">butterflies</span>, kcols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Image classification in progress...</span></code></pre></div>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># Note that we could simply feed the list of images to summary, rather than</span>
<span class="co"># specifying individual images, and they would progress automatically</span>
<span class="co"># with user input.</span>
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">butterflies_class</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="pavo-3-analysing_files/figure-html/unnamed-chunk-55-1.png" alt="The k-means classified images of our butterflies, along with their identified colour palettes" width="480"><p class="caption">
The k-means classified images of our butterflies, along with their identified colour palettes
</p>
</div>
<p>There are other options available, which may be of particular use when the classification process is not producing the results that might be expected. If analysing similar images with homologous colour patterns (e.g. individuals within a species), we could specify a single ‘reference’ image in a list to be initially classified, using <code>refID</code>, with the RGB centres identified in this reference image used as pre-specified cluster centres in subsequent images. This helps to ensure that homologous pattern elements are reliably identified and classified between images. The <code><a href="../reference/classify.html">classify()</a></code> function also offers an interactive option, via <code>interactive = TRUE</code>, wherein we can interactively click on representative colours within an image (or images individually), with the RGB values then used as cluster centres. This is particularly useful when images are being erroneously classified, since the standard k-means process randomly selects initial centres and so may (by chance) miss smaller patches, or confuse similar colours. By specifying the initial RGB centres interactively, it can be easier for the process to distinguish or ‘pull apart’ colours in such circumstances. Ultimately, there are several combinations of options available during classification, which are also documented in the help file. For our butterflies, for example, we could use any of the following (though some would be very ill-advised):</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Automatic classification.</span>
<span class="va">butterflies_class</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/classify.html">classify</a></span><span class="op">(</span><span class="va">butterflies</span>, kcols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Image classification in progress...</span>

<span class="co"># Automatic classification using a reference image, identified by name.</span>
<span class="va">butterflies_class</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/classify.html">classify</a></span><span class="op">(</span><span class="va">butterflies</span>, refID <span class="op">=</span> <span class="st">'h_melpomene'</span>, kcols <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>
<span class="co">#&gt; Image classification in progress...</span></code></pre></div>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Classification using interactively-specified centres for each image, with no</span>
<span class="co"># need to specify kcols (since it will be inferred from the numbers of colours selected)</span>
<span class="va">butterflies_class</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/classify.html">classify</a></span><span class="op">(</span><span class="va">butterflies</span>, interactive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># Classification using interactively-specified centres for the single reference image</span>
<span class="co"># (here, the first in the list). We could also specify reference image using it's name, </span>
<span class="co"># as above.</span>
<span class="va">butterflies_class</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/classify.html">classify</a></span><span class="op">(</span><span class="va">butterflies</span>, refID <span class="op">=</span> <span class="fl">1</span>, interactive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
</div>
<div id="adjacency-and-boundary-strength-analyses-and-overall-pattern-contrasts" class="section level2">
<h2 class="hasAnchor">
<a href="#adjacency-and-boundary-strength-analyses-and-overall-pattern-contrasts" class="anchor"></a>Adjacency and Boundary Strength Analyses, and Overall Pattern Contrasts</h2>
<p>The adjacency <span class="citation">(Endler <a href="#ref-Endler2012" role="doc-biblioref">2012</a>)</span> and boundary strength <span class="citation">(Endler, Cole, and Kranz <a href="#ref-Endler2018" role="doc-biblioref">2018</a>)</span> analyses offer a powerful method for quantifying various features of colour pattern geometry. Briefly, the process entails classifying grid-sampled locations within a visual scene into a number of discrete colour classes. This may be achieved using <code><a href="../reference/classify.html">classify()</a></code> as described above before sampling the now-classified image, and/or by taking grid-sampled reflectance spectra across a visual scene, visually modelling and clustering the results <span class="citation">(as per “method 1” in Endler <a href="#ref-Endler2012" role="doc-biblioref">2012</a>, example below)</span>. The column-wise and row-wise colour-class transitions between adjacent points within this grid are then tallied, and from this a suite of summary statistics on pattern structure — from simple colour proportions, through to colour diversity and pattern complexity — are estimated. If the colour ‘distance’ between adjacent colour classes is known, such as might be estimated using spectral data for receptor-noise modelling via <code><a href="../reference/vismodel.html">vismodel()</a></code> and <code><a href="../reference/coldist.html">coldist()</a></code>, then this can also be incorporated to derive several measures of the salience of patch boundaries <span class="citation">(Endler, Cole, and Kranz <a href="#ref-Endler2018" role="doc-biblioref">2018</a>)</span>. Finally, if estimates of hue, saturation, and/or brightness are known for each discrete colour in a pattern, then these can be included through the <code>hsl</code> argument for general, area-weighted measures of overall pattern contrast <span class="citation">(Endler and Mielke <a href="#ref-Endler2005" role="doc-biblioref">2005</a>)</span>. In <code>pavo</code>, this is all carried out through the <code><a href="../reference/adjacent.html">adjacent()</a></code> function, which takes several arguments that should be well understood.</p>
<div id="example-1-simple-mounted-butterflies" class="section level3">
<h3 class="hasAnchor">
<a href="#example-1-simple-mounted-butterflies" class="anchor"></a>Example 1 (simple): mounted butterflies</h3>
<p>We’ll begin with a simple example using mounted butterflies, by first loading and classifying our images in the manner outlined above. For the adjacency analysis proper, we’ll specify a full x-dimension scale of 200 mm using <code>xscale</code> (which could be ignored if we specify scales using <code><a href="../reference/procimg.html">procimg()</a></code>), and a sampling density of 200, for a 200 x 200 grid. Since the white background of our images is fake and of no interest, we will also exclude it by specifying the colour-class ID that corresponds to it (for more complex backgrounds, we could instead use <code><a href="../reference/procimg.html">procimg()</a></code>, as above, and ignore this argument).</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># Load up our images</span>
<span class="va">butterflies</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getimg.html">getimg</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"testdata/images/"</span>, package <span class="op">=</span> <span class="st">'pavo'</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; 2 files found; importing images.</span>

<span class="co"># Automatically classify discrete colour patches (as confirmed by spectral modelling)</span>
<span class="co"># using k-means clustering, with values of 'k' that have been validated a priori.</span>
<span class="va">butterflies_class</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/classify.html">classify</a></span><span class="op">(</span><span class="va">butterflies</span>, kcols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Image classification in progress...</span>

<span class="co"># Run the adjacency analysis, subsampling the image in a 200x200 grid, and excluding</span>
<span class="co"># the white background.</span>
<span class="va">butterflies_adj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adjacent.html">adjacent</a></span><span class="op">(</span><span class="va">butterflies_class</span>, xscale <span class="op">=</span> <span class="fl">200</span>, xpts <span class="op">=</span> <span class="fl">200</span>, bkgID <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>

<span class="co"># Take a look</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">butterflies_adj</span><span class="op">)</span>
<span class="co">#&gt;             k     N n_off       p_1       p_2       p_3        p_4     q_1_1</span>
<span class="co">#&gt; h_melpomene 4 54064  5111 0.5833824 0.3408824 0.0406250 0.03511029 0.5649970</span>
<span class="co">#&gt; papilio     3 63241  4371 0.1995597 0.5609434 0.2394969         NA 0.1726728</span>
<span class="co">#&gt;                   q_1_2     q_2_2      q_2_3      q_3_3      q_1_4      q_2_4</span>
<span class="co">#&gt; h_melpomene 0.018644569 0.3071175 0.01825614 0.02683856 0.01294762 0.03486608</span>
<span class="co">#&gt; papilio     0.006941699 0.5485682 0.01315602 0.20964248         NA         NA</span>
<span class="co">#&gt;                   q_3_4       q_4_4     t_1_2     t_2_3     t_1_4     t_2_4</span>
<span class="co">#&gt; h_melpomene 0.009821693 0.006510802 0.1972217 0.1931129 0.1369595 0.3688124</span>
<span class="co">#&gt; papilio              NA          NA 0.1004347 0.1903455        NA        NA</span>
<span class="co">#&gt;                 t_3_4      q_1_3     t_1_3          m        m_r        m_c</span>
<span class="co">#&gt; h_melpomene 0.1038936         NA        NA 0.09395221 0.08779412 0.10011029</span>
<span class="co">#&gt; papilio            NA 0.04901883 0.7092199 0.06868910 0.07805031 0.05932789</span>
<span class="co">#&gt;                     A       Sc       St        Jc        Jt         B        Rt</span>
<span class="co">#&gt; h_melpomene 0.8769739 2.176663 4.136277 0.5441657 0.6893795 0.8839757 0.5665327</span>
<span class="co">#&gt; papilio     1.3155754 2.428125 1.820461 0.8093752 0.6068204 0.2150538 0.5493113</span>
<span class="co">#&gt;             Rab m_dS s_dS cv_dS m_dL s_dL cv_dL m_hue s_hue var_hue m_sat s_sat</span>
<span class="co">#&gt; h_melpomene   1   NA   NA    NA   NA   NA    NA    NA    NA      NA    NA    NA</span>
<span class="co">#&gt; papilio       1   NA   NA    NA   NA   NA    NA    NA    NA      NA    NA    NA</span>
<span class="co">#&gt;             cv_sat m_lum s_lum cv_lum</span>
<span class="co">#&gt; h_melpomene     NA    NA    NA     NA</span>
<span class="co">#&gt; papilio         NA    NA    NA     NA</span></code></pre></div>
<p>The output is a suite of statistics describing colour pattern structure and geometry in a tidy data frame that, since we fed it a list of images, contains one row for each of our samples. As always, the function’s help file (<code><a href="../reference/adjacent.html">?adjacent</a></code>) summarises the meaning of each output variable, and the original references listed therein contain the full discussion. Note that some of the output is <code>NA</code>, which might occur when the focal statistic simply makes no sense for the image in question (such as the transition frequency between colour categories 3 and 4, for an image that contains only 3 colour categories). Others may result when we do not give <code><a href="../reference/adjacent.html">adjacent()</a></code> enough information, such as the colour-distances required to calculate the boundary-strength statistics, or the colourmetrics required for overall pattern contrast. If we do give the function this information, we can see that those final variables now contain sensible output.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Create a fake matrix of pairwise colour- and luminance distances between all</span>
<span class="co"># colour patten elements, as might be attained through visual modelling of spectral data.</span>
<span class="va">distances</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>c1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>,
                        c2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span>,
                        dS <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">10.6</span>, <span class="fl">5.1</span>, <span class="fl">4.4</span><span class="op">)</span>,
                        dL <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1.1</span>, <span class="fl">2.5</span>, <span class="fl">3.2</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Take a look</span>
<span class="va">distances</span>
<span class="co">#&gt;   c1 c2   dS  dL</span>
<span class="co">#&gt; 1  1  2 10.6 1.1</span>
<span class="co">#&gt; 2  1  3  5.1 2.5</span>
<span class="co">#&gt; 3  2  3  4.4 3.2</span>

<span class="co"># And our fake hue angles (in radians), saturation, and luminance values, for each</span>
<span class="co"># colour pattern element</span>
<span class="va">hsl_vals</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>patch <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,
                       hue <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1.2</span>, <span class="fl">2.2</span>, <span class="fl">1.6</span><span class="op">)</span>,
                       lum <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">5</span>, <span class="fl">7</span><span class="op">)</span>,
                       sat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3.5</span>, <span class="fl">1.1</span>, <span class="fl">6.3</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Take a look</span>
<span class="va">hsl_vals</span>
<span class="co">#&gt;   patch hue lum sat</span>
<span class="co">#&gt; 1     1 1.2  10 3.5</span>
<span class="co">#&gt; 2     2 2.2   5 1.1</span>
<span class="co">#&gt; 3     3 1.6   7 6.3</span>

<span class="co"># Now feed this information into the adjacency analysis using the 'less-colourful'</span>
<span class="co"># of our two images, for convenience (though this could be readily extended to</span>
<span class="co"># include a list of images along with a list of distances and hsl values)</span>
<span class="fu"><a href="../reference/adjacent.html">adjacent</a></span><span class="op">(</span><span class="va">butterflies_class</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, xscale <span class="op">=</span> <span class="fl">200</span>, xpts <span class="op">=</span> <span class="fl">200</span>, bkgID <span class="op">=</span> <span class="fl">1</span>, coldists <span class="op">=</span> <span class="va">distances</span>, hsl <span class="op">=</span> <span class="va">hsl_vals</span><span class="op">)</span>
<span class="co">#&gt;         k     N n_off       p_1       p_2       p_3     q_1_1       q_1_2</span>
<span class="co">#&gt; papilio 3 63241  4371 0.1995597 0.5609434 0.2394969 0.1726728 0.006941699</span>
<span class="co">#&gt;             q_2_2      q_1_3      q_2_3     q_3_3     t_1_2     t_1_3     t_2_3</span>
<span class="co">#&gt; papilio 0.5485682 0.04901883 0.01315602 0.2096425 0.1004347 0.7092199 0.1903455</span>
<span class="co">#&gt;                 m        m_r        m_c        A       Sc       St        Jc</span>
<span class="co">#&gt; papilio 0.0686891 0.07805031 0.05932789 1.315575 2.428125 1.820461 0.8093752</span>
<span class="co">#&gt;                Jt         B        Rt Rab     m_dS    s_dS     cv_dS     m_dL</span>
<span class="co">#&gt; papilio 0.6068204 0.2150538 0.5493113   1 5.519149 2.56091 0.4640044 2.492633</span>
<span class="co">#&gt;             s_dL     cv_dL    m_hue     s_hue   var_hue    m_sat    s_sat</span>
<span class="co">#&gt; papilio 0.802252 0.3218492 1.663715 0.4154308 0.0826731 2.824327 2.812416</span>
<span class="co">#&gt;            cv_sat    m_lum    s_lum    cv_lum</span>
<span class="co">#&gt; papilio 0.9957825 6.476792 2.530441 0.3906936</span></code></pre></div>
</div>
<div id="example-2-complex-whole-visual-environments" class="section level3">
<h3 class="hasAnchor">
<a href="#example-2-complex-whole-visual-environments" class="anchor"></a>Example 2 (complex): whole visual environments</h3>
<p>Of course colour patterns are not always so simple, and we may want to analyse the spatial geometry of more complex or subtle scenes. In such cases, it may be preferable to largely forego image-based clustering and classification altogether and instead use spectral data directly in an adjacency-style analysis <span class="citation">(as per “method 1” in Endler <a href="#ref-Endler2012" role="doc-biblioref">2012</a>)</span>. One such approach would have us take spectral measurements across a patch in an evenly-spaced grid</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># Load up our image of a camouflaged lizard</span>
<span class="va">lizard</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getimg.html">getimg</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"testdata/images/vig/"</span>, package <span class="op">=</span> <span class="st">'pavo'</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; 1 files found; importing images.</span>

<span class="co"># Take a look at it</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">lizard</span><span class="op">)</span>

<span class="co"># And overlay a simple 'sampling grid' that might guide us as to where we should collect</span>
<span class="co"># our reflectance measurements. Note that the grid here is likely too coarse to capture the necessary detail,</span>
<span class="co"># though the sampling density is something to be closely considered on a case-by-case basis</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">lizard</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/points.html">points</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/expand.grid.html">expand.grid</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">lizard</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="fl">25</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">lizard</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="fl">25</span><span class="op">)</span><span class="op">)</span>, pch <span class="op">=</span> <span class="fl">16</span>, col <span class="op">=</span> <span class="st">'red'</span><span class="op">)</span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="pavo-3-analysing_files/figure-html/unnamed-chunk-60-1.png" alt="A cryptic lizard, along with a possible sampling grid for spectral measurement." width="45%"><p class="caption">
A cryptic lizard, along with a possible sampling grid for spectral measurement.
</p>
</div>
<p>For the sake of this example we’ll simulate 100 reflectance spectra, as though we had collected them following the above grid. We’ll do this by creating four very similar groups of of 25 spectra, which will also allow us to get a sense of how our clustering methods perform. Of course real spectra will be more variable, but the same principles apply.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">12352</span><span class="op">)</span>  <span class="co"># For reprodicubility</span>

<span class="co"># Generate some fake spectra that were grid-sampled across our scene</span>
<span class="va">fakescene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span>
  <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span><span class="op">(</span><span class="va">cbind</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">25</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">dnorm</a></span><span class="op">(</span><span class="fl">300</span><span class="op">:</span><span class="fl">700</span>, <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">320</span>, <span class="fl">330</span><span class="op">)</span>, <span class="fl">20</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,
  <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span><span class="op">(</span><span class="va">cbind</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">25</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">dnorm</a></span><span class="op">(</span><span class="fl">300</span><span class="op">:</span><span class="fl">700</span>, <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">440</span>, <span class="fl">450</span><span class="op">)</span>, <span class="fl">20</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,
  <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span><span class="op">(</span><span class="va">cbind</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">25</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">dnorm</a></span><span class="op">(</span><span class="fl">300</span><span class="op">:</span><span class="fl">700</span>, <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">550</span>, <span class="fl">560</span><span class="op">)</span>, <span class="fl">20</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,
  <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span><span class="op">(</span><span class="va">cbind</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">25</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">dnorm</a></span><span class="op">(</span><span class="fl">300</span><span class="op">:</span><span class="fl">700</span>, <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">650</span>, <span class="fl">660</span><span class="op">)</span>, <span class="fl">20</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
  <span class="op">)</span>

<span class="co"># Convert them to rspce objects</span>
<span class="va">fakescene</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as.rspec.html">as.rspec</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>wl <span class="op">=</span> <span class="fl">300</span><span class="op">:</span><span class="fl">700</span>, <span class="va">fakescene</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Now instead of calling <code>cluster()</code> on the image and arbitrarily selecting <code>kcols</code> (which would likely be quite inaccurate due to the complexity of the scene), we can visually model the spectra, calculate the colour-distances between them, and cluster the results into a data-image that can be fed directly into <code><a href="../reference/adjacent.html">adjacent()</a></code>. Below we’ll use a centroid-based ‘hartigan’ clustering method via the package <code>NbClust</code>, but there are numerous possible approaches that might be used, which should be closely considered with reference to the original methods’ publications.</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Visually model our spectra in a tetrahedral model of the blue tit, specifying</span>
<span class="co"># relative = FALSE so that we can estimate noise-calibrated distances.</span>
<span class="va">vis.fakescene</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vismodel.html">vismodel</a></span><span class="op">(</span><span class="va">fakescene</span>, visual <span class="op">=</span> <span class="st">'bluetit'</span>, relative <span class="op">=</span> <span class="cn">FALSE</span>, scale <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span>

<span class="co"># Calculate noise-weighted distances, before converting them to xyz coordinates</span>
<span class="co"># in noise-corrected colourspace.</span>
<span class="va">jnd.fakescene</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/jnd2xyz.html">jnd2xyz</a></span><span class="op">(</span><span class="fu"><a href="../reference/coldist.html">coldist</a></span><span class="op">(</span><span class="va">vis.fakescene</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Create a distance matrix, for clustering</span>
<span class="va">jnd.mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span><span class="op">(</span><span class="va">jnd.fakescene</span>, method <span class="op">=</span> <span class="st">'euclidean'</span>, diag <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co"># Zero all distance &lt; 1 ('just noticeable distance', in this model), which corresponds</span>
<span class="co"># to a theoretical threshold of discrimination in the receptor-noise limited</span>
<span class="co"># model.</span>
<span class="va">jnd.mat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">jnd.mat</span> <span class="op">&lt;</span> <span class="fl">1</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span>

<span class="co"># Load up a library and use k-means clustering to estimate the number of discrete</span>
<span class="co"># colours present in our sample.</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://sites.google.com/site/malikacharrad/research/nbclust-package">NbClust</a></span><span class="op">)</span>
<span class="va">clust</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/NbClust/man/NbClust.html">NbClust</a></span><span class="op">(</span><span class="va">jnd.fakescene</span>, diss <span class="op">=</span> <span class="va">jnd.mat</span>, distance <span class="op">=</span> <span class="cn">NULL</span>, method <span class="op">=</span> <span class="st">'centroid'</span>, index <span class="op">=</span> <span class="st">'hartigan'</span><span class="op">)</span>
<span class="va">clust</span><span class="op">$</span><span class="va">Best.nc</span>
<span class="co">#&gt; Number_clusters     Value_Index </span>
<span class="co">#&gt;           4.000        1639.353</span></code></pre></div>
<p>This approach has identified four distinct colour patches, as simulated. To be able to feed this information directly in to <code><a href="../reference/adjacent.html">adjacent()</a></code> we need to arrange the now-classified spectral data into a xyz data-image matrix, in which <code>x</code> and <code>y</code> coordinates specify the location of the sample in our scene, and the z values at each location specify the colour-class, or group, into which each sample has been classified (an integer). Many clustering packages offers the data’s classification as output in some form, so we can simply rearrange it into our original 10 x 10 sampling matrix before passing it to <code><a href="../reference/adjacent.html">adjacent()</a></code>.</p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Rearrange the data into a colour-classified image matrix, and take a look at it.</span>
<span class="co"># Note that is the same structure as the output of 'classify()'.</span>
<span class="va">mat.fakescene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="va">clust</span><span class="op">$</span><span class="va">Best.partition</span><span class="op">)</span><span class="op">)</span>, <span class="fl">10</span>, <span class="fl">10</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">mat.fakescene</span><span class="op">)</span>
<span class="co">#&gt;      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]</span>
<span class="co">#&gt; [1,]    1    1    1    2    2    3    3    3    4     4</span>
<span class="co">#&gt; [2,]    1    1    1    2    2    3    3    3    4     4</span>
<span class="co">#&gt; [3,]    1    1    1    2    2    3    3    3    4     4</span>
<span class="co">#&gt; [4,]    1    1    1    2    2    3    3    3    4     4</span>
<span class="co">#&gt; [5,]    1    1    1    2    2    3    3    3    4     4</span>
<span class="co">#&gt; [6,]    1    1    2    2    2    3    3    4    4     4</span>

<span class="co"># Run the adjacency analysis</span>
<span class="fu"><a href="../reference/adjacent.html">adjacent</a></span><span class="op">(</span><span class="va">mat.fakescene</span>, xscale <span class="op">=</span> <span class="fl">200</span><span class="op">)</span>
<span class="co">#&gt;     k   N n_off  p_1  p_2  p_3  p_4     q_1_1      q_1_2     q_2_2      q_2_3</span>
<span class="co">#&gt; img 4 180    32 0.25 0.25 0.25 0.25 0.2055556 0.06111111 0.2055556 0.05555556</span>
<span class="co">#&gt;         q_3_3      q_3_4     q_4_4   t_1_2  t_2_3   t_3_4     m   m_r   m_c  A</span>
<span class="co">#&gt; img 0.2055556 0.06111111 0.2055556 0.34375 0.3125 0.34375 0.008 0.015 0.001 15</span>
<span class="co">#&gt;     Sc       St Jc        Jt  B Rt Rab m_dS s_dS cv_dS m_dL s_dL cv_dL m_hue</span>
<span class="co">#&gt; img  4 2.994152  1 0.4990253 NA NA  NA   NA   NA    NA   NA   NA    NA    NA</span>
<span class="co">#&gt;     s_hue var_hue m_sat s_sat cv_sat m_lum s_lum cv_lum</span>
<span class="co">#&gt; img    NA      NA    NA    NA     NA    NA    NA     NA</span></code></pre></div>
<p>The output reflects several aspects of our simulated dataset, which is reassuring. There are four colour classes (<code>k</code>= 4) or equal proportion (<code>p_x</code>), 180 transitions in total (<code>N</code>, 9 transitions per row/column x 10 rows/columns), and the colour diversity (<code>Sc</code>) equals <code>k</code>, and relative colour diversity (<code>Jc</code>) equals 1, since all colours occur in equal proportions.</p>
</div>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-Andersson2006">
<p>Andersson, Staffan, and Maria Prager. 2006. “Quantification of Coloration.” In <em>Bird Coloration, Volume 1: Mechanisms and Measurements</em>, edited by Geoffrey E. Hill and Kevin J. McGraw, 1:640. Bird Coloration. Harvard University Press.</p>
</div>
<div id="ref-Backhaus1991">
<p>Backhaus, Werner. 1991. “Color Opponent Coding in the Visual System of the Honeybee.” <em>Vision Research</em> 31 (7): 1381–97. <a href="https://doi.org/10.1016/0042-6989(91)90059-E">https://doi.org/10.1016/0042-6989(91)90059-E</a>.</p>
</div>
<div id="ref-Briscoe2001">
<p>Briscoe, Adriana D., and Lars Chittka. 2001. “The Evolution of Color Vision in Insects.” <em>Annual Review of Entomology</em> 46 (1): 471–510. <a href="https://doi.org/10.1146/annurev.ento.46.1.471">https://doi.org/10.1146/annurev.ento.46.1.471</a>.</p>
</div>
<div id="ref-Chittka1992">
<p>Chittka, Lars. 1992. “The Colour Hexagon: A Chromaticity Diagram Based on Photoreceptor Excitations as a Generalized Representation of Colour Opponency.” <em>Journal of Comparative Physiology A</em> 170 (5): 533–43. <a href="https://doi.org/10.1007/BF00199331">https://doi.org/10.1007/BF00199331</a>.</p>
</div>
<div id="ref-Endler1990">
<p>Endler, John A. 1990. “On the Measurement and Classification of Colour in Studies of Animal Colour Patterns.” <em>Biological Journal of the Linnean Society</em> 41 (4): 315–52. <a href="https://doi.org/10.1111/j.1095-8312.1990.tb00839.x">https://doi.org/10.1111/j.1095-8312.1990.tb00839.x</a>.</p>
</div>
<div id="ref-Endler2012">
<p>———. 2012. “A Framework for Analysing Colour Pattern Geometry: Adjacent Colours.” <em>Biological Journal of the Linnean Society</em> 107 (2): 233–53. <a href="https://doi.org/10.1111/j.1095-8312.2012.01937.x">https://doi.org/10.1111/j.1095-8312.2012.01937.x</a>.</p>
</div>
<div id="ref-Endler2018">
<p>Endler, John A., Gemma L. Cole, and Alexandrea M. Kranz. 2018. “Boundary Strength Analysis: Combining Colour Pattern Geometry and Coloured Patch Visual Properties for Use in Predicting Behaviour and Fitness.” Edited by Simon Blomberg. <em>Methods in Ecology and Evolution</em>, August. <a href="https://doi.org/10.1111/2041-210X.13073">https://doi.org/10.1111/2041-210X.13073</a>.</p>
</div>
<div id="ref-Endler2005">
<p>Endler, John A., and Paul W. Mielke. 2005. “Comparing Entire Colour Patterns as Birds See Them.” <em>Biological Journal of the Linnean Society</em> 86 (4): 405–31. <a href="https://doi.org/10.1111/j.1095-8312.2005.00540.x">https://doi.org/10.1111/j.1095-8312.2005.00540.x</a>.</p>
</div>
<div id="ref-Goldsmith1990">
<p>Goldsmith, Timothy H. 1990. “Optimization, Constraint, and History in the Evolution of Eyes.” <em>The Quarterly Review of Biology</em> 65 (3): 281–322. <a href="https://doi.org/10.1086/416840">https://doi.org/10.1086/416840</a>.</p>
</div>
<div id="ref-Govardovskii2000">
<p>Govardovskii, Victor I., Nanna Fyhrquist, Tom Reuter, Dmitry G. Kuzmin, and Kristian Donner. 2000. “In Search of the Visual Pigment Template.” <em>Visual Neuroscience</em> 17 (4): 509–28. <a href="https://doi.org/10.1017/S0952523800174036">https://doi.org/10.1017/S0952523800174036</a>.</p>
</div>
<div id="ref-Hart2005">
<p>Hart, Nathan S., and Misha Vorobyev. 2005. “Modelling Oil Droplet Absorption Spectra and Spectral Sensitivities of Bird Cone Photoreceptors.” <em>Journal of Comparative Physiology A</em> 191 (4): 381–92. <a href="https://doi.org/10.1007/s00359-004-0595-3">https://doi.org/10.1007/s00359-004-0595-3</a>.</p>
</div>
<div id="ref-Maia2018">
<p>Maia, Rafael, and Thomas E White. 2018. “Comparing Colors Using Visual Models.” <em>Behavioral Ecology</em> 29 (3): 649–59. <a href="https://doi.org/10.1093/beheco/ary017">https://doi.org/10.1093/beheco/ary017</a>.</p>
</div>
<div id="ref-Montgomerie2006">
<p>Montgomerie, Robert. 2006. “Analyzing Colors.” In <em>Bird Coloration, Volume 1: Mechanisms and Measurements</em>, edited by Geoffrey E. Hill and Kevin J. McGraw, 1:640. Bird Coloration. Harvard University Press.</p>
</div>
<div id="ref-Renoult2017">
<p>Renoult, Julien P., Almut Kelber, and H. Martin Schaefer. 2017. “Colour Spaces in Ecology and Evolutionary Biology.” <em>Biological Reviews</em> 92 (1): 292–315. <a href="https://doi.org/10.1111/brv.12230">https://doi.org/10.1111/brv.12230</a>.</p>
</div>
<div id="ref-Stoddard2008">
<p>Stoddard, Mary Caswell, and Richard O. Prum. 2008. “Evolution of Avian Plumage Color in a Tetrahedral Color Space: A Phylogenetic Analysis of New World Buntings.” <em>The American Naturalist</em> 171 (6): 755–76. <a href="https://doi.org/10.1086/587526">https://doi.org/10.1086/587526</a>.</p>
</div>
<div id="ref-Stoddard2011">
<p>Stoddard, Mary Caswell, and Martin Stevens. 2011. “Avian Vision and the Evolution of Egg Color Mimicry in the Common Cuckoo.” <em>Evolution</em> 65 (7): 2004–13. <a href="https://doi.org/10.1111/j.1558-5646.2011.01262.x">https://doi.org/10.1111/j.1558-5646.2011.01262.x</a>.</p>
</div>
<div id="ref-Troje1993">
<p>Troje, Nikolaus. 1993. “Spectral Categories in the Learning Behaviour of Blowflies.” <em>Zeitschrift Für Naturforschung C</em> 48 (1-2): 96–104.</p>
</div>
<div id="ref-Vorobyev2001">
<p>Vorobyev, Misha, Robert Brandt, Dagmar Peitsch, Simon B Laughlin, and Randolf Menzel. 2001. “Colour Thresholds and Receptor Noise: Behaviour and Physiology Compared.” <em>Vision Research</em> 41 (5): 639–53. <a href="https://doi.org/10.1016/S0042-6989(00)00288-1">https://doi.org/10.1016/S0042-6989(00)00288-1</a>.</p>
</div>
<div id="ref-Vorobyev1998">
<p>Vorobyev, Misha, Daniel Colaco Osorio, Andrew T. D. Bennett, N. Justin Marshall, and Innes C. Cuthill. 1998. “Tetrachromacy, Oil Droplets and Bird Plumage Colours.” <em>Journal of Comparative Physiology A</em> 183 (5): 621–33. <a href="https://doi.org/10.1007/s003590050286">https://doi.org/10.1007/s003590050286</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Thomas White, Rafael Maia, Hugo Gruson, John Endler, Chad Eliason, Pierre-Paul Bitton.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '1988255e7e235678bc3c7f1b86b9ecc5',
    indexName: 'rafaelmaia_pavo',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
